<!DOCTYPE html>
<!-- saved from url=(0056)https://cole-trapnell-lab.github.io/cicero-release/docs/ -->
<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
   
   <meta http-equiv="X-UA-Compatible" content="IE=edge">
   <meta name="viewport" content="width=device-width, initial-scale=1">

   <title>Cicero</title>
   <meta name="description" content="Cicero - Predicting the cis-regulatory landscape">

   <link rel="shortcut icon" type="image/png" href="https://cole-trapnell-lab.github.io/cicero-release/images/favicon.png">

  <link rel="stylesheet" href="./Cicero_files/main.css">
  <link rel="stylesheet" href="./Cicero_files/base16-atelierlakeside.light.css">
  <link rel="stylesheet" href="./Cicero_files/bootstrap-toc.min.css">

  <link rel="canonical" href="https://cole-trapnell-lab.github.io/cicero-release/docs/">

<script async="" src="./Cicero_files/js"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-123277373-1');
</script>

<style type="text/css">.MathJax_Hover_Frame {border-radius: .25em; -webkit-border-radius: .25em; -moz-border-radius: .25em; -khtml-border-radius: .25em; box-shadow: 0px 0px 15px #83A; -webkit-box-shadow: 0px 0px 15px #83A; -moz-box-shadow: 0px 0px 15px #83A; -khtml-box-shadow: 0px 0px 15px #83A; border: 1px solid #A6D ! important; display: inline-block; position: absolute}
.MathJax_Menu_Button .MathJax_Hover_Arrow {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 4px; -webkit-border-radius: 4px; -moz-border-radius: 4px; -khtml-border-radius: 4px; font-family: 'Courier New',Courier; font-size: 9px; color: #F0F0F0}
.MathJax_Menu_Button .MathJax_Hover_Arrow span {display: block; background-color: #AAA; border: 1px solid; border-radius: 3px; line-height: 0; padding: 4px}
.MathJax_Hover_Arrow:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_Hover_Arrow:hover span {background-color: #CCC!important}
</style><style type="text/css">#MathJax_About {position: fixed; left: 50%; width: auto; text-align: center; border: 3px outset; padding: 1em 2em; background-color: #DDDDDD; color: black; cursor: default; font-family: message-box; font-size: 120%; font-style: normal; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 15px; -webkit-border-radius: 15px; -moz-border-radius: 15px; -khtml-border-radius: 15px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_About.MathJax_MousePost {outline: none}
.MathJax_Menu {position: absolute; background-color: white; color: black; width: auto; padding: 5px 0px; border: 1px solid #CCCCCC; margin: 0; cursor: default; font: menu; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 5px; -webkit-border-radius: 5px; -moz-border-radius: 5px; -khtml-border-radius: 5px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
.MathJax_MenuItem {padding: 1px 2em; background: transparent}
.MathJax_MenuArrow {position: absolute; right: .5em; padding-top: .25em; color: #666666; font-size: .75em}
.MathJax_MenuActive .MathJax_MenuArrow {color: white}
.MathJax_MenuArrow.RTL {left: .5em; right: auto}
.MathJax_MenuCheck {position: absolute; left: .7em}
.MathJax_MenuCheck.RTL {right: .7em; left: auto}
.MathJax_MenuRadioCheck {position: absolute; left: .7em}
.MathJax_MenuRadioCheck.RTL {right: .7em; left: auto}
.MathJax_MenuLabel {padding: 1px 2em 3px 1.33em; font-style: italic}
.MathJax_MenuRule {border-top: 1px solid #DDDDDD; margin: 4px 3px}
.MathJax_MenuDisabled {color: GrayText}
.MathJax_MenuActive {background-color: #606872; color: white}
.MathJax_MenuDisabled:focus, .MathJax_MenuLabel:focus {background-color: #E8E8E8}
.MathJax_ContextMenu:focus {outline: none}
.MathJax_ContextMenu .MathJax_MenuItem:focus {outline: none}
#MathJax_AboutClose {top: .2em; right: .2em}
.MathJax_Menu .MathJax_MenuClose {top: -10px; left: -10px}
.MathJax_MenuClose {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; font-family: 'Courier New',Courier; font-size: 24px; color: #F0F0F0}
.MathJax_MenuClose span {display: block; background-color: #AAA; border: 1.5px solid; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; line-height: 0; padding: 8px 0 6px}
.MathJax_MenuClose:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_MenuClose:hover span {background-color: #CCC!important}
.MathJax_MenuClose:hover:focus {outline: none}
</style><style type="text/css">.MathJax_Preview .MJXf-math {color: inherit!important}
</style><style type="text/css">.MJX_Assistive_MathML {position: absolute!important; top: 0; left: 0; clip: rect(1px, 1px, 1px, 1px); padding: 1px 0 0 0!important; border: 0!important; height: 1px!important; width: 1px!important; overflow: hidden!important; display: block!important; -webkit-touch-callout: none; -webkit-user-select: none; -khtml-user-select: none; -moz-user-select: none; -ms-user-select: none; user-select: none}
.MJX_Assistive_MathML.MJX_Assistive_MathML_Block {width: 100%!important}
</style><style type="text/css">#MathJax_Zoom {position: absolute; background-color: #F0F0F0; overflow: auto; display: block; z-index: 301; padding: .5em; border: 1px solid black; margin: 0; font-weight: normal; font-style: normal; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; -webkit-box-sizing: content-box; -moz-box-sizing: content-box; box-sizing: content-box; box-shadow: 5px 5px 15px #AAAAAA; -webkit-box-shadow: 5px 5px 15px #AAAAAA; -moz-box-shadow: 5px 5px 15px #AAAAAA; -khtml-box-shadow: 5px 5px 15px #AAAAAA; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_ZoomOverlay {position: absolute; left: 0; top: 0; z-index: 300; display: inline-block; width: 100%; height: 100%; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
#MathJax_ZoomFrame {position: relative; display: inline-block; height: 0; width: 0}
#MathJax_ZoomEventTrap {position: absolute; left: 0; top: 0; z-index: 302; display: inline-block; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
</style><style type="text/css">.MathJax_Preview {color: #888}
#MathJax_Message {position: fixed; left: 1em; bottom: 1.5em; background-color: #E6E6E6; border: 1px solid #959595; margin: 0px; padding: 2px 8px; z-index: 102; color: black; font-size: 80%; width: auto; white-space: nowrap}
#MathJax_MSIE_Frame {position: absolute; top: 0; left: 0; width: 0px; z-index: 101; border: 0px; margin: 0px; padding: 0px}
.MathJax_Error {color: #CC0000; font-style: italic}
</style><style type="text/css">.MJXp-script {font-size: .8em}
.MJXp-right {-webkit-transform-origin: right; -moz-transform-origin: right; -ms-transform-origin: right; -o-transform-origin: right; transform-origin: right}
.MJXp-bold {font-weight: bold}
.MJXp-italic {font-style: italic}
.MJXp-scr {font-family: MathJax_Script,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-frak {font-family: MathJax_Fraktur,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-sf {font-family: MathJax_SansSerif,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-cal {font-family: MathJax_Caligraphic,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-mono {font-family: MathJax_Typewriter,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-largeop {font-size: 150%}
.MJXp-largeop.MJXp-int {vertical-align: -.2em}
.MJXp-math {display: inline-block; line-height: 1.2; text-indent: 0; font-family: 'Times New Roman',Times,STIXGeneral,serif; white-space: nowrap; border-collapse: collapse}
.MJXp-display {display: block; text-align: center; margin: 1em 0}
.MJXp-math span {display: inline-block}
.MJXp-box {display: block!important; text-align: center}
.MJXp-box:after {content: " "}
.MJXp-rule {display: block!important; margin-top: .1em}
.MJXp-char {display: block!important}
.MJXp-mo {margin: 0 .15em}
.MJXp-mfrac {margin: 0 .125em; vertical-align: .25em}
.MJXp-denom {display: inline-table!important; width: 100%}
.MJXp-denom > * {display: table-row!important}
.MJXp-surd {vertical-align: top}
.MJXp-surd > * {display: block!important}
.MJXp-script-box > *  {display: table!important; height: 50%}
.MJXp-script-box > * > * {display: table-cell!important; vertical-align: top}
.MJXp-script-box > *:last-child > * {vertical-align: bottom}
.MJXp-script-box > * > * > * {display: block!important}
.MJXp-mphantom {visibility: hidden}
.MJXp-munderover {display: inline-table!important}
.MJXp-over {display: inline-block!important; text-align: center}
.MJXp-over > * {display: block!important}
.MJXp-munderover > * {display: table-row!important}
.MJXp-mtable {vertical-align: .25em; margin: 0 .125em}
.MJXp-mtable > * {display: inline-table!important; vertical-align: middle}
.MJXp-mtr {display: table-row!important}
.MJXp-mtd {display: table-cell!important; text-align: center; padding: .5em 0 0 .5em}
.MJXp-mtr > .MJXp-mtd:first-child {padding-left: 0}
.MJXp-mtr:first-child > .MJXp-mtd {padding-top: 0}
.MJXp-mlabeledtr {display: table-row!important}
.MJXp-mlabeledtr > .MJXp-mtd:first-child {padding-left: 0}
.MJXp-mlabeledtr:first-child > .MJXp-mtd {padding-top: 0}
.MJXp-merror {background-color: #FFFF88; color: #CC0000; border: 1px solid #CC0000; padding: 1px 3px; font-style: normal; font-size: 90%}
.MJXp-scale0 {-webkit-transform: scaleX(.0); -moz-transform: scaleX(.0); -ms-transform: scaleX(.0); -o-transform: scaleX(.0); transform: scaleX(.0)}
.MJXp-scale1 {-webkit-transform: scaleX(.1); -moz-transform: scaleX(.1); -ms-transform: scaleX(.1); -o-transform: scaleX(.1); transform: scaleX(.1)}
.MJXp-scale2 {-webkit-transform: scaleX(.2); -moz-transform: scaleX(.2); -ms-transform: scaleX(.2); -o-transform: scaleX(.2); transform: scaleX(.2)}
.MJXp-scale3 {-webkit-transform: scaleX(.3); -moz-transform: scaleX(.3); -ms-transform: scaleX(.3); -o-transform: scaleX(.3); transform: scaleX(.3)}
.MJXp-scale4 {-webkit-transform: scaleX(.4); -moz-transform: scaleX(.4); -ms-transform: scaleX(.4); -o-transform: scaleX(.4); transform: scaleX(.4)}
.MJXp-scale5 {-webkit-transform: scaleX(.5); -moz-transform: scaleX(.5); -ms-transform: scaleX(.5); -o-transform: scaleX(.5); transform: scaleX(.5)}
.MJXp-scale6 {-webkit-transform: scaleX(.6); -moz-transform: scaleX(.6); -ms-transform: scaleX(.6); -o-transform: scaleX(.6); transform: scaleX(.6)}
.MJXp-scale7 {-webkit-transform: scaleX(.7); -moz-transform: scaleX(.7); -ms-transform: scaleX(.7); -o-transform: scaleX(.7); transform: scaleX(.7)}
.MJXp-scale8 {-webkit-transform: scaleX(.8); -moz-transform: scaleX(.8); -ms-transform: scaleX(.8); -o-transform: scaleX(.8); transform: scaleX(.8)}
.MJXp-scale9 {-webkit-transform: scaleX(.9); -moz-transform: scaleX(.9); -ms-transform: scaleX(.9); -o-transform: scaleX(.9); transform: scaleX(.9)}
.MathJax_PHTML .noError {vertical-align: ; font-size: 90%; text-align: left; color: black; padding: 1px 3px; border: 1px solid}
</style></head>


  <body data-spy="scroll" data-target="#toc"><div id="MathJax_Message" style="display: none;"></div>
      <div class="container-fullwidth">
  <nav class="navbar navbar-inverse navbar-static-top">
    <div class="container">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="https://cole-trapnell-lab.github.io/cicero-release/">Cicero</a>
      </div>
      <div id="navbar" class="collapse navbar-collapse">
        <ul class="nav navbar-nav">
          <li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/">Documentation</a></li>
          <li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs_m3/">Documentation for Monocle 3</a></li>
          <li><a href="https://cole-trapnell-lab.github.io/cicero-release/papers/">Publications</a></li>
          <li><a href="https://cole-trapnell-lab.github.io/cicero-release/data/">Data</a></li>
          <li><a href="https://github.com/cole-trapnell-lab/cicero-release">GitHub</a></li>
        </ul>
      </div>
    </div>
  </nav>
  </div>



    
    <div class="container">
  <div class="row">
    <div class="col-sm-4">
      <nav id="toc" data-spy="affix" data-toggle="toc" class="affix"><ul class="nav"><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#abstract">Abstract</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#introduction">Introduction</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#installing-cicero"> Installing Cicero </a><ul class="nav"><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#install-from-bioconductor"> Install from Bioconductor </a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#install-from-github"> Install from Github </a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#load-cicero"> Load Cicero </a></li></ul></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#getting-help"> Getting help </a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#recommended-analysis-protocol">Recommended analysis protocol</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#loading-your-data">Loading your data</a><ul class="nav"><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#the-celldataset-class">The CellDataSet class</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#loading-data-from-a-simple-sparse-matrix-format">Loading data from a simple sparse matrix format</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#loading-10x-scatac-seq-data">Loading 10X scATAC-seq data</a></li></ul></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#constructing-cis-regulatory-networks">Constructing cis-regulatory networks</a><ul class="nav"><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#running-cicero">Running Cicero</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#visualizing-cicero-connections">Visualizing Cicero Connections</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#comparing-cicero-connections-to-other-datasets">Comparing Cicero connections to other datasets</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#finding-cis-co-accessibility-networks-ccans-">Finding cis-Co-accessibility Networks (CCANS)</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#cicero-gene-activity-scores">Cicero gene activity scores</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#advanced-visualizaton">Advanced visualizaton</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#important-considerations-for-non-human-data">Important Considerations for Non-Human Data</a></li></ul></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#single-cell-accessibility-trajectories">Single-cell accessibility trajectories</a><ul class="nav"><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#constructing-trajectories-with-accessibility-data">Constructing trajectories with accessibility data</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#differential-accessibility-analysis">Differential Accessibility Analysis</a></li></ul></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#useful-functions">Useful Functions</a><ul class="nav"><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#annotate-cds-by-site">annotate_cds_by_site</a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#find-overlapping-coordinates">find_overlapping_coordinates</a></li></ul></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#frequently-asked-questions">Frequently Asked Questions </a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#citation">Citation </a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#acknowledgements">Acknowledgements </a></li><li><a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#references">References </a></li></ul></nav>
    </div>
    <div class="col-sm-8">
      <h1>Cicero for Monocle/Monocle 2</h1>
      <div class="panel panel-warning"><div class="panel-body">If you are using Monocle 3, please see the documentation <a href="https://cole-trapnell-lab.github.io/cicero-release/docs_m3/">here</a></div></div>
      <h2 id="abstract">Abstract</h2>
      <p> 
        <strong>Cicero</strong> is an R package that provides tools for analyzing single-cell chromatin accessibility 
        experiments. The main function of Cicero is to use single-cell chromatin accessibility data to predict 
        <strong><i>cis</i>-regulatory interactions</strong> (such as those between enhancers and promoters) in the genome by 
        examining co-accessibility. In addition, Cicero extends <a href="http://cole-trapnell-lab.github.io/monocle-release/">Monocle</a> to allow clustering, ordering, and 
        differential accessibility analysis of single cells using chromatin accessibility. For more information about 
        <strong>Cicero</strong>, explore <a href="https://cole-trapnell-lab.github.io/cicero-release/papers/">our publications</a>.
      </p>
    
      <h2 id="introduction">Introduction</h2>

      <p>
        The main purpose of Cicero is to use single-cell chromatin accessibility data to predict regions of the 
        genome that are more likely to be in physical proximity in the nucleus. This can be used to identify putative 
        enhancer-promoter pairs, and to get a sense of the overall stucture of the <i>cis</i>-architecture of a genomic 
        region.
      </p>

      <p>
        Because of the sparsity of single-cell data, cell must be aggregated by similarity to allow robust correction 
        for various technical factors in the data.
      </p>

      <p>
        Ultimately, Cicero provides a "Cicero co-accessibility" score between -1 and 1 between each pair of 
        accessibile peaks within a user defined distance where a higher number indicates higher co-accessibility.
      </p>

      <p> 
        In addition, the Cicero package provides an extension toolkit for analyzing single-cell ATAC-seq experiments 
        using the framework provided by <a href="http://cole-trapnell-lab.github.io/monocle-release/">Monocle</a>. 
        This vignette provides an overview of a single-cell ATAC-Seq analysis workflow with Cicero. For further 
        information and more options, please see the manual pages for the Cicero R package, and 
        <a href="https://cole-trapnell-lab.github.io/cicero-release/papers/">our publications</a>.
      </p>
    
      <p>
        Cicero can help you perform two main types of analysis:
        </p><ul>
          <li><strong>Constructing and analysing <i>cis</i>-regulatory networks.</strong> Cicero analyzes co-accessibility 
            to identify putative <i>cis</i>-regulatory interactions, and uses various techniques to visualize and analyze 
            them.
          </li><li><strong>General single-cell chromatin accessibility analysis .</strong> Cicero also extends the 
            software package <a href="http://cole-trapnell-lab.github.io/monocle-release/">Monocle</a> to allow for 
            identification of differential accessibility, clustering, visualization, and trajectory reconstruction 
            using single-cell chromatin accessibility data.
        </li></ul>
      <p></p>
      <p>
        Before we look at Cicero's functions for each of these analysis tasks, let's see how to install Cicero.
      </p>
    
      <h2 id="installing-cicero"> Installing Cicero </h2>
      <p>
        Cicero runs in the <a href="http://www.r-project.org/">R statistical computing environment</a>. You will 
        need R version 3.4 or higher and Cicero 1.0.0 or higher to have access to the 
        latest features.
      </p>
      <h3 id="install-from-bioconductor"> Install from Bioconductor </h3>
      <p>
        Cicero was recently added to Bioconductor. It has been released as part of Bioconductor 3.8 and 
        can be installed with the following code:
      </p>
<figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span><span class="token function">requireNamespace</span><span class="token punctuation">(</span><span class="token string">"BiocManager"</span><span class="token punctuation">,</span> <span class="token variable">quietly</span> <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
  <span class="token function">install.packages</span><span class="token punctuation">(</span><span class="token string">"BiocManager"</span><span class="token punctuation">)</span>
<span class="token namespace">BiocManager</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">install</span><span class="token punctuation">(</span><span class="token string">"cicero"</span><span class="token punctuation">)</span></code></pre></figure>

      <h3 id="install-from-github"> Install from Github </h3>
      <p> 
        For the absolute latest updates to the Cicero package, you can install Cicero directly from the github repository 
        using these instructions:
      </p>  
      <p>
        Cicero builds upon a package called <a href="http://cole-trapnell-lab.github.io/monocle-release/">Monocle</a>. 
        Before installing Cicero, first, follow 
        <a href="http://cole-trapnell-lab.github.io/monocle-release/docs/#installing-monocle" target="_blank">these instructions</a> 
        to install Monocle. 
      </p>

    <p>
      In addition to Monocle, Cicero requires the following packages from Bioconductor
    </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span><span class="token function">requireNamespace</span><span class="token punctuation">(</span><span class="token string">"BiocManager"</span><span class="token punctuation">,</span> <span class="token variable">quietly</span> <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
    <span class="token function">install.packages</span><span class="token punctuation">(</span><span class="token string">"BiocManager"</span><span class="token punctuation">)</span>
<span class="token namespace">BiocManager</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">install</span><span class="token punctuation">(</span><span class="token function">c</span><span class="token punctuation">(</span><span class="token string">"Gviz"</span><span class="token punctuation">,</span> <span class="token string">"GenomicRanges"</span><span class="token punctuation">,</span> <span class="token string">"rtracklayer"</span><span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre></figure>
    <p>
      You're now ready to install Cicero:
      </p>
    

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">install.packages</span><span class="token punctuation">(</span><span class="token string">"devtools"</span><span class="token punctuation">)</span>
<span class="token namespace">devtools</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">install_github</span><span class="token punctuation">(</span><span class="token string">"cole-trapnell-lab/cicero-release"</span><span class="token punctuation">)</span></code></pre></figure>
<h3 id="load-cicero"> Load Cicero </h3>
      <p>
        After installation, test that Cicero installed correctly by opening a new R session and typing:
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">library</span><span class="token punctuation">(</span><span class="token variable">cicero</span><span class="token punctuation">)</span></code></pre></figure>
      <h2 id="getting-help"> Getting help </h2>

      <p>
        Questions about Cicero should be posted on our 
        <a href="https://groups.google.com/forum/#!forum/cicero-users">Google Group</a>. Please do not email 
        technical questions to Cicero contributors directly.
      </p>

      <h2 id="recommended-analysis-protocol">Recommended analysis protocol</h2>

      <p>
        The Cicero workflow is broken up into broad steps. When there's more than one way to do a certain step, we've 
        labeled the options as follows:
      </p>
      <div class="table-wrapper">
      <table class="table">
        <tbody>
          <tr>
            <td><span class="label label-danger">Required</span></td>
            <td>You need to do this.</td>
          </tr>
          <tr>
            <td><span class="label label-info">Recommended</span> </td>
            <td>Of the ways you could do this, we recommend you try this one first.</td>
          </tr>
          <tr>
            <td><span class="label label-success">Alternative</span></td>
            <td>Of the ways you could do this, this way might work better than the one we usually recommend.</td>
          </tr>
        </tbody>
      </table>
      </div>

      <h2 id="loading-your-data">Loading your data</h2>
      <h3 id="the-celldataset-class">The CellDataSet class</h3>
      <p>
        Cicero holds data in objects of the <code><a href="https://rdrr.io/bioc/monocle/man/CellDataSet.html">
          CellDataSet (CDS)</a></code> class. The class is derived from the Bioconductor <code>ExpressionSet</code> 
        class, which provides a common interface familiar to those who have analyzed microarray experiments with 
        Bioconductor. Monocle provides detailed documentation about how to generate an input CDS
        <a href="http://cole-trapnell-lab.github.io/monocle-release/docs/#the-celldataset-class">here</a>.
      </p>

      <p>
        To modify the CDS object to hold chromatin accessibility rather than expression data, Cicero uses peaks as 
        its feature data <code>fData</code> rather than genes or transcripts. Specifically, many Cicero functions 
        require peak information in the form chr1_10390134_10391134. For example, an input fData table might look 
        like this:

        </p><div class="table-wrapper">
        <table class="table">
          <thead>
            <tr>
              <th></th>
              <th>site_name</th>
              <th>chromosome</th>
              <th>bp1</th>
              <th>bp2</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th>chr10_100002625_100002940</th>
              <th>chr10_100002625_100002940</th>
              <th>10</th>
              <th>100002625</th>
              <th>100002940</th>
            </tr>
            <tr>
              <th>chr10_100006458_100007593</th>
              <th>chr10_100006458_100007593</th>
              <th>10</th>
              <th>100006458</th>
              <th>100007593</th>
            </tr>
            <tr>
              <th>chr10_100011280_100011780</th>
              <th>chr10_100011280_100011780</th>
              <th>10</th>
              <th>100011280</th>
              <th>100011780</th>
            </tr>
            <tr>
              <th>chr10_100013372_100013596</th>
              <th>chr10_100013372_100013596</th>
              <th>10</th>
              <th>100013372</th>
              <th>100013596</th>
            </tr>
            <tr>
              <th>chr10_100015079_100015428</th>
              <th>chr10_100015079_100015428</th>
              <th>10</th>
              <th>100015079</th>
              <th>100015428</th>
            </tr>
          </tbody>
        </table>
        </div>
      <p></p>

      <h3 id="loading-data-from-a-simple-sparse-matrix-format">Loading data from a simple sparse matrix format</h3>
      <p>
        The Cicero package includes a small dataset called <code>cicero_data</code> as an example.

        </p><figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">data</span><span class="token punctuation">(</span><span class="token variable">cicero_data</span><span class="token punctuation">)</span></code></pre></figure>

      <p></p>
      <p>
        For convenience, Cicero includes a function called <code>make_atac_cds</code>. This function takes as input a 
        <code>data.frame</code> or a path to a file in a sparse matrix format. Specifically, this file should be a 
        tab-delimited text file with three columns. The first column is the peak coordinates in the form 
        "chr10_100013372_100013596", the second column is the cell name, and the third column is an integer that 
        represents the number of reads from that cell overlapping that peak. The file should not have a header line. 
      </p>
      <p>
        For example:
      </p> 

        <div class="table-wrapper">
        <table class="table">
          <tbody>
            <tr>
              <th>chr10_100002625_100002940</th>
              <th>cell1</th>
              <th>1</th>
            </tr>
            <tr>
              <th>chr10_100006458_100007593</th>
              <th>cell2</th>
              <th>2</th>
            </tr>
            <tr>
              <th>chr10_100006458_100007593</th>
              <th>cell3</th>
              <th>1</th>
            </tr>
            <tr>
              <th>chr10_100013372_100013596</th>
              <th>cell2</th>
              <th>1</th>
            </tr>
            <tr>
              <th>chr10_100015079_100015428</th>
              <th>cell4</th>
              <th>3</th>
            </tr>
          </tbody>
        </table>
        </div>
      <p></p>

      <p>
        The output of <code>make_atac_cds</code> is a valid CDS object ready to be input into downstream Cicero 
        functions.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">input_cds</span> <span class="token operator">&lt;-</span> <span class="token function">make_atac_cds</span><span class="token punctuation">(</span><span class="token variable">cicero_data</span><span class="token punctuation">,</span> <span class="token variable">binarize</span> <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span></code></pre></figure>

      <div class="panel panel-info">
          <div class="panel-heading">
            <h4 class="panel-title">make_atac_cds</h4>
          </div>
          <div class="panel-body">
            <table class="table">
              <thead>
                <tr>
                  <th>Option</th>
                  <th>Default</th>
                  <th>Notes</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td><code>binarize</code></td>
                  <td>FALSE</td>
                  <td>Logical, should the count matrix be converted to binary? Because of the sparsity of single-cell 
                    data, we don't generally expect more than 1 read per cell per peak. We have found that converting 
                    a count matrix to a binary "is this site open in this cell" can ultimately give clearer results.
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      <h3 id="loading-10x-scatac-seq-data">Loading 10X scATAC-seq data</h3>
      <p>
        If your scATAC-seq data comes from the 10x Genomics platform, here is an example of creating your input CDS 
        from the output of Cell Ranger ATAC. The output of interest will be in the folder called 
        <strong>filtered_peak_bc_matrix</strong>.
      </p>
<figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token comment" spellcheck="true"># read in matrix data using the Matrix package</span>
<span class="token variable">indata</span> <span class="token operator">&lt;-</span> <span class="token namespace">Matrix</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">readMM</span><span class="token punctuation">(</span><span class="token string">"filtered_peak_bc_matrix/matrix.mtx"</span><span class="token punctuation">)</span> 
<span class="token comment" spellcheck="true"># binarize the matrix</span>
<span class="token variable">indata</span><span class="token punctuation">@</span><span class="token property">x</span><span class="token punctuation">[</span><span class="token variable">indata</span><span class="token punctuation">@</span><span class="token property">x</span> <span class="token operator">&gt;</span> <span class="token number">0</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token number">1</span>

<span class="token comment" spellcheck="true"># format cell info</span>
<span class="token variable">cellinfo</span> <span class="token operator">&lt;-</span> <span class="token function">read.table</span><span class="token punctuation">(</span><span class="token string">"filtered_peak_bc_matrix/barcodes.tsv"</span><span class="token punctuation">)</span>
<span class="token function">row.names</span><span class="token punctuation">(</span><span class="token variable">cellinfo</span><span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token variable">cellinfo</span><span class="token punctuation">$</span><span class="token property">V1</span>
<span class="token function">names</span><span class="token punctuation">(</span><span class="token variable">cellinfo</span><span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token string">"cells"</span>

<span class="token comment" spellcheck="true"># format peak info</span>
<span class="token variable">peakinfo</span> <span class="token operator">&lt;-</span> <span class="token function">read.table</span><span class="token punctuation">(</span><span class="token string">"filtered_peak_bc_matrix/peaks.bed"</span><span class="token punctuation">)</span>
<span class="token function">names</span><span class="token punctuation">(</span><span class="token variable">peakinfo</span><span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token string">"chr"</span><span class="token punctuation">,</span> <span class="token string">"bp1"</span><span class="token punctuation">,</span> <span class="token string">"bp2"</span><span class="token punctuation">)</span>
<span class="token variable">peakinfo</span><span class="token punctuation">$</span><span class="token property">site_name</span> <span class="token operator">&lt;-</span> <span class="token function">paste</span><span class="token punctuation">(</span><span class="token variable">peakinfo</span><span class="token punctuation">$</span><span class="token property">chr</span><span class="token punctuation">,</span> <span class="token variable">peakinfo</span><span class="token punctuation">$</span><span class="token property">bp1</span><span class="token punctuation">,</span> <span class="token variable">peakinfo</span><span class="token punctuation">$</span><span class="token property">bp2</span><span class="token punctuation">,</span> <span class="token variable">sep</span><span class="token operator">=</span><span class="token string">"_"</span><span class="token punctuation">)</span>
<span class="token function">row.names</span><span class="token punctuation">(</span><span class="token variable">peakinfo</span><span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token variable">peakinfo</span><span class="token punctuation">$</span><span class="token property">site_name</span>

<span class="token function">row.names</span><span class="token punctuation">(</span><span class="token variable">indata</span><span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token function">row.names</span><span class="token punctuation">(</span><span class="token variable">peakinfo</span><span class="token punctuation">)</span>
<span class="token function">colnames</span><span class="token punctuation">(</span><span class="token variable">indata</span><span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token function">row.names</span><span class="token punctuation">(</span><span class="token variable">cellinfo</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># make CDS</span>
<span class="token variable">fd</span> <span class="token operator">&lt;-</span> <span class="token namespace">methods</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">new</span><span class="token punctuation">(</span><span class="token string">"AnnotatedDataFrame"</span><span class="token punctuation">,</span> <span class="token variable">data</span> <span class="token operator">=</span> <span class="token variable">peakinfo</span><span class="token punctuation">)</span>
<span class="token variable">pd</span> <span class="token operator">&lt;-</span> <span class="token namespace">methods</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">new</span><span class="token punctuation">(</span><span class="token string">"AnnotatedDataFrame"</span><span class="token punctuation">,</span> <span class="token variable">data</span> <span class="token operator">=</span> <span class="token variable">cellinfo</span><span class="token punctuation">)</span>
<span class="token variable">input_cds</span> <span class="token operator">&lt;-</span>  <span class="token function">suppressWarnings</span><span class="token punctuation">(</span><span class="token function">newCellDataSet</span><span class="token punctuation">(</span><span class="token variable">indata</span><span class="token punctuation">,</span>
                            <span class="token variable">phenoData</span> <span class="token operator">=</span> <span class="token variable">pd</span><span class="token punctuation">,</span>
                            <span class="token variable">featureData</span> <span class="token operator">=</span> <span class="token variable">fd</span><span class="token punctuation">,</span>
                            <span class="token variable">expressionFamily</span><span class="token operator">=</span><span class="token namespace">VGAM</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">binomialff</span><span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
                            <span class="token variable">lowerDetectionLimit</span><span class="token operator">=</span><span class="token number">0</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token variable">input_cds</span><span class="token punctuation">@</span><span class="token property">expressionFamily</span><span class="token punctuation">@</span><span class="token property">vfamily</span> <span class="token operator">&lt;-</span> <span class="token string">"binomialff"</span>
<span class="token variable">input_cds</span> <span class="token operator">&lt;-</span> <span class="token namespace">monocle</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">detectGenes</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true">#Ensure there are no peaks included with zero reads</span>
<span class="token variable">input_cds</span> <span class="token operator">&lt;-</span> <span class="token variable">input_cds</span><span class="token punctuation">[</span><span class="token namespace">Matrix</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">rowSums</span><span class="token punctuation">(</span><span class="token function">exprs</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token operator">!</span><span class="token operator">=</span> <span class="token number">0</span><span class="token punctuation">,</span><span class="token punctuation">]</span> </code></pre></figure>

      <h2 id="constructing-cis-regulatory-networks">Constructing <i>cis</i>-regulatory networks</h2>

      <h3 id="running-cicero">Running Cicero</h3>
      <h4 data-toc-text="Create a Cicero CDS">Create a Cicero CDS 
        <span class="label label-danger">Required</span></h4>

      <p>
        Because single-cell chromatin accessibility data is extremely sparse, accurate estimation of co-accessibility 
        scores requires us to aggregate similar cells to create more dense count data. Cicero does this using a 
        k-nearest-neighbors approach which creates overlapping sets of cells. Cicero constructs these sets based on a 
        reduced dimension coordinate map of cell similarity, for example, from a tSNE or DDRTree map.
      </p>

      <p>
        You can use any dimensionality reduction method to base your aggregated CDS on. We will show you how to 
        create two versions, tSNE and DDRTree (below). Both of these dimensionality reduction methods are available 
        from <a href="http://cole-trapnell-lab.github.io/monocle-release/">Monocle</a> (and loaded by Cicero).
      </p>
    
      <p>
        Once you have your reduced dimension coordinate map, you can use the function <code>make_cicero_cds</code> to 
        create your aggregated CDS object. The input to <code>make_cicero_cds</code> is your input CDS object, and 
        your reduced dimension coordinate map. The reduced dimension map <code>reduced_coordinates</code> should be 
        in the form of a <code>data.frame</code> or a <code>matrix</code> where the row names match the cell IDs from 
        the <code>pData</code> table of your CDS. The columns of <code>reduced_coordinates</code> should be the 
        coordinates of the reduced dimension object, for example:
      </p>
      <div class="table-wrapper">
      <table class="table" style="overflow:auto;">
        <tbody>
          <tr>
            <th></th>
            <th>ddrtree_coord1</th>
            <th>ddrtree_coord2</th>
          </tr>
          <tr>
            <th>cell1</th>
            <th>-0.7084047</th>
            <th>-0.7232994</th>
          </tr>
          <tr>
            <th>cell2</th>
            <th>-4.4767964</th>
            <th>0.8237284</th>
          </tr>
          <tr>
            <th>cell3</th>
            <th>1.4870098</th>
            <th>-0.4723493</th>
          </tr>
        </tbody>
      </table>
      </div>

      <p>
        Here is an example of both dimensionality reduction and creation of a Cicero CDS. Using monocle as a guide, 
        we first find tSNE coordinates for our input_cds:
        </p><figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">set.seed</span><span class="token punctuation">(</span><span class="token number">2017</span><span class="token punctuation">)</span>
<span class="token variable">input_cds</span> <span class="token operator">&lt;-</span> <span class="token function">detectGenes</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span>
<span class="token variable">input_cds</span> <span class="token operator">&lt;-</span> <span class="token function">estimateSizeFactors</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># *** if you are using Monocle 3 alpha, you need to run the following line as well!</span>
<span class="token comment" spellcheck="true"># NOTE: Cicero does not yet support the Monocle 3 beta release (monocle3 package). We hope</span>
<span class="token comment" spellcheck="true"># to update soon!</span>
<span class="token comment" spellcheck="true">#input_cds &lt;- preprocessCDS(input_cds, norm_method = "none")</span>
<span class="token variable">input_cds</span> <span class="token operator">&lt;-</span> <span class="token function">reduceDimension</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">,</span> <span class="token variable">max_components</span> <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">,</span> <span class="token variable">num_dim</span><span class="token operator">=</span><span class="token number">6</span><span class="token punctuation">,</span>
                      <span class="token variable">reduction_method</span> <span class="token operator">=</span> <span class="token string">'tSNE'</span><span class="token punctuation">,</span> <span class="token variable">norm_method</span> <span class="token operator">=</span> <span class="token string">"none"</span><span class="token punctuation">)</span></code></pre></figure>

        For more information on the above code, see the Monocle website on
        <a href="http://cole-trapnell-lab.github.io/monocle-release/docs/#clustering-cells">clustering cells</a>
      <p></p>

      <p>
        Next, we access the tSNE coordinates from the input CDS object where they are stored by Monocle and run 
        <code>make_cicero_cds</code>:
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">tsne_coords</span> <span class="token operator">&lt;-</span> <span class="token function">t</span><span class="token punctuation">(</span><span class="token function">reducedDimA</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token function">row.names</span><span class="token punctuation">(</span><span class="token variable">tsne_coords</span><span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token function">row.names</span><span class="token punctuation">(</span><span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token variable">cicero_cds</span> <span class="token operator">&lt;-</span> <span class="token function">make_cicero_cds</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">,</span> <span class="token variable">reduced_coordinates</span> <span class="token operator">=</span> <span class="token variable">tsne_coords</span><span class="token punctuation">)</span></code></pre></figure>

      <h4 data-toc-text="Run Cicero">Run Cicero <span class="label label-danger">Required</span></h4>

      <p>
        The main function of the Cicero package is to estimate the co-accessiblity of sites in the genome in order to 
        predict <i>cis</i>-regulatory interactions. There are two ways to get this information:
        </p><ul>
          <li><strong><code>run_cicero</code>: get Cicero outputs with all defaults</strong> The function 
            <code>run_cicero</code> will call each of the relevant pieces of Cicero code using default values, and 
            calculating best-estimate parameters as it goes. For most users, this will be the best place to start.
          </li><li><strong>Call functions separately, for more flexibility</strong> For users wanting more flexibility in 
            the parameters that are called, and those that want access to intermediate information, Cicero allows you 
            to call each of the component parts separately.
        </li></ul>
      <p></p>
    
      <div class="panel panel-warning">
        <div class="panel-heading">
          <h4 class="panel-title" data-toc-skip="">Consider non-default parameters</h4>
        </div>
        <div class="panel-body">
          The default parameters are designed for data from human and mouse cells. There are certain non-default 
          parameters that each user should consider, especially users not using data from humans. Read the section 
          <a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#important-considerations-for-non-human-data">Important Considerations 
            for Non-Human Data</a> for more information.
        </div>
      </div>

      <h4 data-toc-text="run_cicero"><code>run_cicero</code> <span class="label label-info">Recommended</span></h4>

      <p>
        The easiest way to get Cicero co-accessibility scores is to run <code>run_cicero</code>. To run 
        <code>run_cicero</code>, you need a cicero CDS object (created above) and a genome coordinates file, which 
        contains the lengths of each of the chromosomes in your organism. The human hg19 coordinates are included 
        with the package and can be accessed with <code>data("human.hg19.genome")</code>. Here is an example call, 
        continuing with our example data:
      </p>
    
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">data</span><span class="token punctuation">(</span><span class="token string">"human.hg19.genome"</span><span class="token punctuation">)</span>
<span class="token variable">sample_genome</span> <span class="token operator">&lt;-</span> <span class="token function">subset</span><span class="token punctuation">(</span><span class="token variable">human</span><span class="token punctuation">.</span><span class="token variable">hg19</span><span class="token punctuation">.</span><span class="token variable">genome</span><span class="token punctuation">,</span> <span class="token variable">V1</span> <span class="token operator">==</span> <span class="token string">"chr18"</span><span class="token punctuation">)</span>
<span class="token variable">conns</span> <span class="token operator">&lt;-</span> <span class="token function">run_cicero</span><span class="token punctuation">(</span><span class="token variable">cicero_cds</span><span class="token punctuation">,</span> <span class="token variable">sample_genome</span><span class="token punctuation">)</span> <span class="token comment" spellcheck="true"># Takes a few minutes to run</span>
<span class="token function">head</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">)</span></code></pre></figure>

      <div class="table-wrapper">
      <table class="table" style="overflow:auto;">
        <tbody>
          <tr>
            <th>Peak1</th>
            <th>Peak2</th>
            <th>coaccess</th>
          </tr>
          <tr>
            <th>chr18_10025_10225</th>
            <th>chr18_104385_104585</th>
            <th>0.03791526</th>
          </tr>
          <tr>
            <th>chr18_10025_10225</th>
            <th>chr18_10603_11103</th>
            <th>0.86971957</th>
          </tr>
          <tr>
            <th>chr18_10025_10225</th>
            <th>chr18_111867_112367</th>
            <th>0.00000000</th>
          </tr>
        </tbody>
      </table>
      </div>

      <h4 data-toc-text="Call Cicero functions individually">Call Cicero functions individually 
        <span class="label label-success">Alternative</span></h4>

      <p> 
        The alternative to calling <code>run_cicero</code> is to run each piece of the Cicero pipeline in pieces. The 
        three functions that make up the Cicero pipeline are:
        </p><ul>
          <li><strong><code>estimate_distance_parameter</code></strong>. This function calculates the distance 
            penalty parameter based on small random windows of the genome.
          </li><li><strong><code>generate_cicero_models</code></strong>. This function uses the distance parameter 
            determined above and uses graphical LASSO to calculate the co-accessibility scores of overlapping windows 
            of the genome using a distance-based penalty.
          </li><li><strong><code>assemble_connections</code></strong>. This function takes as input the output of 
            <code>generate_cicero_models</code> and reconciles the overlapping models to create the final list of 
            co-accessibility scores.
        </li></ul>
      <p></p>

      <p>
        Use the manual pages in R (example: <code>?estimate_distance_parameter</code>) to see all options. Key 
        options are detailed below:
      </p>

      <div class="panel panel-info">
        <div class="panel-heading">
          <h4 class="panel-title">estimate_distance_parameter</h4>
        </div>
        <div class="panel-body">
          <table class="table">
            <thead>
              <tr>
                <th>Option</th>
                <th>Default</th>
                <th>Notes</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td><code>window</code></td>
                <td>500000</td>
                <td>The window parameter controls how large of a window (in base pairs) in the genome is used for 
                  calculating each individual model. This parameter will control the furthest distance between sites 
                  that are compared. This parameter <strong>must</strong> be the same as the window parameter for 
                  <code>generate_cicero_models</code>.</td>
              </tr>
              <tr>
                <td><code>sample_num</code></td>
                <td>100</td>
                <td>This is the number of sample regions to calculate a distance parameter for.</td>
              </tr>
              <tr>
                <td><code>distance_constraint</code></td>
                <td>250000</td>
                <td>The <code>distance_constraint</code> controls the distance at which Cicero expects few meaningful 
                  <i>cis</i>-regulatory contacts. <code>estimate_distance_parameter</code> uses this value to estimate the 
                  distance parameter by increasing regularization until there are few contacts at a distance beyond 
                  the <code>distance_constraint</code>. The <code>distance_constraint</code> <strong>must</strong> be 
                  sufficiently lower than the window parameter - as a rule of thumb, set your window to be at least 
                  1/3 larger than your <code>distance_constraint</code>.</td>
              </tr>
              <tr>
                <td><code>s</code></td>
                <td>0.75</td>
                <td><code>s</code> is a constant that captures the power-law distribution of contact frequencies 
                  between different locations in the genome as a function of their linear distance. For a complete 
                  discussion of the various polymer models of DNA packed into the nucleus and of justifiable values 
                  for <code>s</code>, we refer readers to (Dekker et al., 2013). We use a value of 0.75 by default in 
                  Cicero, which corresponds to the tension globule polymer model of DNA (Sanborn et al., 2015). For 
                  data not from humans, we recommend you read further discussion of this parameter here: 
                  <a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#important-considerations-for-non-human-data">Important 
                    Considerations for Non-Human Data</a>. This parameter <strong>must</strong> be the same as the 
                  <code>s</code> parameter for <code>generate_cicero_models</code>.</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>

      <div class="panel panel-info">
        <div class="panel-heading">
          <h4 class="panel-title">generate_cicero_models</h4>
        </div>
        <div class="panel-body">
          <table class="table">
            <thead>
              <tr>
                <th>Option</th>
                <th>Default</th>
                <th>Notes</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td><code>window</code></td>
                <td>500000</td>
                <td>The window parameter controls how large of a window (in base pairs) in the genome is used for 
                    calculating each individual model. This parameter will control the furthest distance between sites 
                    that are compared. This parameter <strong>must</strong> be the same as the window parameter for 
                    <code>estimate_distance_parameter</code>.</td>
              </tr>
              <tr>
                <td><code>distance_parameter</code></td>
                <td>100</td>
                <td>The <code>distance_parameter</code> controls the distance-based scaling of the regularization 
                  parameter in graphical LASSO. This parameter is generally the mean of the sample values calculated 
                  using <code>estimate_distance_parameter</code></td>
              </tr>
              <tr>
                <td><code>s</code></td>
                <td>0.75</td>
                <td>See above in the documentation for <code>estimate_distance_parameter</code>. This parameter 
                  <strong>must</strong> be the same as the <code>s</code> parameter for 
                  <code>estimate_distance_parameter</code>.</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>

      <div class="panel panel-info">
        <div class="panel-heading">
          <h4 class="panel-title">assemble_connections</h4>
        </div>
        <div class="panel-body">
          <table class="table">
            <thead>
              <tr>
                <th>Option</th>
                <th>Default</th>
                <th>Notes</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td><code>silent</code></td>
                <td>FALSE</td>
                <td>Logical, whether to print summary information.</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>

      <h3 id="visualizing-cicero-connections">Visualizing Cicero Connections</h3>

      <p>
        The Cicero package includes a general plotting function for visualizing co-accessibility called
        <code>plot_connections</code>. This function uses the 
        <a href="https://bioconductor.org/packages/release/bioc/html/Gviz.html">Gviz</a> framework for plotting 
        genome browser-style plots. We have adapted a function from the
          <a href="https://bioconductor.org/packages/release/bioc/html/Sushi.html">Sushi</a> R package for 
          mapping connections. <code>plot_connections</code> has many options, some detailed in the 
        <a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#advanced-visualizaton">Advanced Visualization</a> section, but to get a 
        basic plot from your co-accessibility table is quite simple:
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">data</span><span class="token punctuation">(</span><span class="token variable">gene_annotation_sample</span><span class="token punctuation">)</span>
<span class="token function">plot_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token number">8575097</span><span class="token punctuation">,</span> <span class="token number">8839855</span><span class="token punctuation">,</span> 
                <span class="token variable">gene_model</span> <span class="token operator">=</span> <span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> 
                <span class="token variable">coaccess_cutoff</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">25</span><span class="token punctuation">,</span> 
                <span class="token variable">connection_width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span> 
                <span class="token variable">collapseTranscripts</span> <span class="token operator">=</span> <span class="token string">"longest"</span> <span class="token punctuation">)</span></code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/Cicero_example.png">
      </div>
      <h3 id="comparing-cicero-connections-to-other-datasets">Comparing Cicero connections to other datasets</h3>
      <p>
        Often, it is useful to compare Cicero connections to other datasets with similar kinds of connections. For 
        example, you might want to compare the output of Cicero to ChIA-PET ligations. To do this, Cicero includes a 
        function called <code>compare_connections</code>. This function takes as input two data frames of connection 
        pairs, <code>conns1</code> and <code>conns2</code>, and returns a logical vector of connections from 
        <code>conns1</code> found in <code>conns2</code>. The comparison in this function is conducted using the 
        <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a> 
        package, and uses the <code>max_gap</code> argument from that package to allow slop in the comparisons.
      </p>

      <p>
        For example, if we wanted to compare our Cicero predictions to a set of (made-up) ChIA-PET connections, we 
        could run:
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">chia_conns</span> <span class="token operator">&lt;-</span>  <span class="token function">data.frame</span><span class="token punctuation">(</span><span class="token variable">Peak1</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token string">"chr18_10000_10200"</span><span class="token punctuation">,</span> <span class="token string">"chr18_10000_10200"</span><span class="token punctuation">,</span> 
                                  <span class="token string">"chr18_49500_49600"</span><span class="token punctuation">)</span><span class="token punctuation">,</span> 
                        <span class="token variable">Peak2</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token string">"chr18_10600_10700"</span><span class="token punctuation">,</span> <span class="token string">"chr18_111700_111800"</span><span class="token punctuation">,</span> 
                                  <span class="token string">"chr18_10600_10700"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token function">head</span><span class="token punctuation">(</span><span class="token variable">chia_conns</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true">#               Peak1               Peak2</span>
<span class="token comment" spellcheck="true"># 1 chr18_10000_10200   chr18_10600_10700</span>
<span class="token comment" spellcheck="true"># 2 chr18_10000_10200 chr18_111700_111800</span>
<span class="token comment" spellcheck="true"># 3 chr18_49500_49600   chr18_10600_10700</span>

<span class="token variable">conns</span><span class="token punctuation">$</span><span class="token property">in_chia</span> <span class="token operator">&lt;-</span> <span class="token function">compare_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> <span class="token variable">chia_conns</span><span class="token punctuation">)</span>

<span class="token function">head</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true">#               Peak1               Peak2    coaccess in_chia</span>
<span class="token comment" spellcheck="true"># 2 chr18_10025_10225   chr18_10603_11103  0.85209126    TRUE</span>
<span class="token comment" spellcheck="true"># 3 chr18_10025_10225   chr18_11604_13986 -0.55433268   FALSE</span>
<span class="token comment" spellcheck="true"># 4 chr18_10025_10225   chr18_49557_50057 -0.43594546   FALSE</span>
<span class="token comment" spellcheck="true"># 5 chr18_10025_10225   chr18_50240_50740 -0.43662436   FALSE</span>
<span class="token comment" spellcheck="true"># 6 chr18_10025_10225 chr18_104385_104585  0.00000000   FALSE</span>
<span class="token comment" spellcheck="true"># 7 chr18_10025_10225 chr18_111867_112367  0.01405174   FALSE</span></code></pre></figure>
      <p>
        You may find that this overlap is too strict when comparing completely distinct datasets. Looking carefully, 
        the 2nd line of the ChIA-PET data matches fairly closely to the last line shown of conns. The difference is 
        only ~67 base pairs, which could be a matter of peak-calling. This is where the <code>max_gap</code> 
        parameter can be useful:
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">conns</span><span class="token punctuation">$</span><span class="token property">in_chia_100</span> <span class="token operator">&lt;-</span> <span class="token function">compare_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> <span class="token variable">chia_conns</span><span class="token punctuation">,</span> <span class="token variable">maxgap</span><span class="token operator">=</span><span class="token number">100</span><span class="token punctuation">)</span>

<span class="token function">head</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true">#               Peak1               Peak2    coaccess in_chia in_chia_100</span>
<span class="token comment" spellcheck="true"># 2 chr18_10025_10225   chr18_10603_11103  0.85209126    TRUE        TRUE</span>
<span class="token comment" spellcheck="true"># 3 chr18_10025_10225   chr18_11604_13986 -0.55433268   FALSE       FALSE</span>
<span class="token comment" spellcheck="true"># 4 chr18_10025_10225   chr18_49557_50057 -0.43594546   FALSE       FALSE</span>
<span class="token comment" spellcheck="true"># 5 chr18_10025_10225   chr18_50240_50740 -0.43662436   FALSE       FALSE</span>
<span class="token comment" spellcheck="true"># 6 chr18_10025_10225 chr18_104385_104585  0.00000000   FALSE       FALSE</span>
<span class="token comment" spellcheck="true"># 7 chr18_10025_10225 chr18_111867_112367  0.01405174   FALSE        TRUE</span></code></pre></figure>

      <p>
        In addition, Cicero's plotting function has a way to compare datasets visually. To do this, use the 
        <code>comparison_track</code> argument. The comparison data frame <strong>must</strong> include a third 
        columns beyond the first two peak columns called "coaccess". This is how the plotting function determines the 
        height of the plotted connections. This could be a quantitative measure, like the number of ligations in 
        ChIA-PET, or simply a column of 1s. More info on plotting options in manual pages 
        <code>?plot_connections</code> and in the <a href="https://cole-trapnell-lab.github.io/cicero-release/docs/#advanced-visualizaton">Advanced 
          Visualization</a> section. 
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token comment" spellcheck="true"># Add a column of 1s called "coaccess"</span>
<span class="token variable">chia_conns</span> <span class="token operator">&lt;-</span>  <span class="token function">data.frame</span><span class="token punctuation">(</span><span class="token variable">Peak1</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token string">"chr18_10000_10200"</span><span class="token punctuation">,</span> <span class="token string">"chr18_10000_10200"</span><span class="token punctuation">,</span> 
                                  <span class="token string">"chr18_49500_49600"</span><span class="token punctuation">)</span><span class="token punctuation">,</span> 
                        <span class="token variable">Peak2</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token string">"chr18_10600_10700"</span><span class="token punctuation">,</span> <span class="token string">"chr18_111700_111800"</span><span class="token punctuation">,</span> 
                                  <span class="token string">"chr18_10600_10700"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
                        <span class="token variable">coaccess</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span> <span class="token number">1</span><span class="token punctuation">,</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token function">plot_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token number">10000</span><span class="token punctuation">,</span> <span class="token number">112367</span><span class="token punctuation">,</span> 
                <span class="token variable">gene_model</span> <span class="token operator">=</span> <span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> 
                <span class="token variable">coaccess_cutoff</span> <span class="token operator">=</span> <span class="token number">0</span><span class="token punctuation">,</span>
                <span class="token variable">connection_width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_track</span> <span class="token operator">=</span> <span class="token variable">chia_conns</span><span class="token punctuation">,</span>
                <span class="token variable">include_axis_track</span> <span class="token operator">=</span> <span class="token boolean">F</span><span class="token punctuation">,</span>
                <span class="token variable">collapseTranscripts</span> <span class="token operator">=</span> <span class="token string">"longest"</span><span class="token punctuation">)</span> </code></pre></figure>
      <div class="text-center">
        <img src="./Cicero_files/comparison_track.png">
      </div>
      <h3 id="finding-cis-co-accessibility-networks-ccans-">Finding <i>cis</i>-Co-accessibility Networks (CCANS)</h3>
      <p>
        In addition to pairwise co-accessibility scores, Cicero also has a function to find <i>Cis</i>-Co-accessibility 
        Networks (CCANs), which are modules of sites that are highly co-accessible with one another. We use the 
        Louvain community detection algorithm (Blondel et al., 2008) to find clusters of sites that tended to be 
        co-accessible. The function <code>generate_ccans</code> takes as input a connection data frame and outputs a 
        data frame with CCAN assignments for each input peak. Sites not included in the output data frame were not 
        assigned a CCAN.
      </p>
      <br>
      <br>
      <div class="text-center">
        <img src="./Cicero_files/ccan_diagram.png">
      </div>
      <br>
      <br>
      <p>
        The function <code>generate_ccans</code> has one optional input called <code>coaccess_cutoff_override</code>. 
        When <code>coaccess_cutoff_override</code> is NULL, the function will determine and report an appropriate 
        co-accessibility score cutoff value for CCAN generation based on the number of overall CCANs at varying 
        cutoffs. You can also set <code>coaccess_cutoff_override</code> to be a numeric between 0 and 1, to override 
        the cutoff-finding part of the function. This option is useful if you feel that the cutoff found 
        automatically was too strict or loose, or for speed if you are rerunning the code and know what the cutoff 
        will be, since the cutoff finding procedure can be slow.
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">CCAN_assigns</span> <span class="token operator">&lt;-</span> <span class="token function">generate_ccans</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># [1] "Co-accessibility cutoff used: 0.24"</span>

<span class="token function">head</span><span class="token punctuation">(</span><span class="token variable">CCAN_assigns</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true">#                                    Peak CCAN</span>
<span class="token comment" spellcheck="true"># chr18_10025_10225     chr18_10025_10225    1</span>
<span class="token comment" spellcheck="true"># chr18_10603_11103     chr18_10603_11103    1</span>
<span class="token comment" spellcheck="true"># chr18_11604_13986     chr18_11604_13986    1</span>
<span class="token comment" spellcheck="true"># chr18_49557_50057     chr18_49557_50057    1</span>
<span class="token comment" spellcheck="true"># chr18_50240_50740     chr18_50240_50740    1</span>
<span class="token comment" spellcheck="true"># chr18_157883_158536 chr18_157883_158536    1</span></code></pre></figure>

      <h3 id="cicero-gene-activity-scores">Cicero gene activity scores</h3>
      <p>
        We have found that often, accessibility at promoters is a poor predictor of gene expression. However, using 
        Cicero links, we are able to get a better sense of the overall accessibility of a promoter and it's 
        associated distal sites. This combined score of regional accessibility has a better concordance with gene 
        expression. We call this score the Cicero gene activity score, and it is calculated using two functions.
      </p>
      <p>
        The initial function is called <code>build_gene_activity_matrix</code>. This function takes an input CDS and 
        a Cicero connection list, and outputs an unnormalized table of gene activity scores. 
        <strong>IMPORTANT</strong>: the input CDS must have a column in the fData table called "gene" which indicates 
        the gene if that peak is a promoter, and <code>NA</code> if the peak is distal. One way to add this column is 
        demonstrated below.
      </p>
        
      <p>  
        The output of <code>build_gene_activity_matrix</code> is <strong>unnormalized</strong>. It must be normalized 
        using a second function called <code>normalize_gene_activities</code>. If you intend to compare gene 
        activities across different datasets of subsets of data, then all gene activity subsets should be normalized 
        together, by passing in a list of unnormalized matrices. If you only wish to normalized one matrix, simply 
        pass it to the function on its own. <code>normalize_gene_activities</code> also requires a named vector of 
        of total accessible sites per cell. This is easily found in the pData table of your CDS, called 
        "num_genes_expressed". See below for an example. Normalized gene activity scores range from 0 to 1.
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token comment" spellcheck="true">#### Add a column for the fData table indicating the gene if a peak is a promoter ####</span>
<span class="token comment" spellcheck="true"># Create a gene annotation set that only marks the transcription start sites of </span>
<span class="token comment" spellcheck="true"># the genes. We use this as a proxy for promoters.</span>
<span class="token comment" spellcheck="true"># To do this we need the first exon of each transcript</span>
<span class="token variable">pos</span> <span class="token operator">&lt;-</span> <span class="token function">subset</span><span class="token punctuation">(</span><span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> <span class="token variable">strand</span> <span class="token operator">==</span> <span class="token string">"+"</span><span class="token punctuation">)</span>
<span class="token variable">pos</span> <span class="token operator">&lt;-</span> <span class="token variable">pos</span><span class="token punctuation">[</span><span class="token function">order</span><span class="token punctuation">(</span><span class="token variable">pos</span><span class="token punctuation">$</span><span class="token property">start</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span> 
<span class="token variable">pos</span> <span class="token operator">&lt;-</span> <span class="token variable">pos</span><span class="token punctuation">[</span><span class="token operator">!</span><span class="token function">duplicated</span><span class="token punctuation">(</span><span class="token variable">pos</span><span class="token punctuation">$</span><span class="token property">transcript</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span> <span class="token comment" spellcheck="true"># remove all but the first exons per transcript</span>
<span class="token variable">pos</span><span class="token punctuation">$</span><span class="token property">end</span> <span class="token operator">&lt;-</span> <span class="token variable">pos</span><span class="token punctuation">$</span><span class="token property">start</span> <span class="token operator">+</span> <span class="token number">1</span> <span class="token comment" spellcheck="true"># make a 1 base pair marker of the TSS</span>

<span class="token variable">neg</span> <span class="token operator">&lt;-</span> <span class="token function">subset</span><span class="token punctuation">(</span><span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> <span class="token variable">strand</span> <span class="token operator">==</span> <span class="token string">"-"</span><span class="token punctuation">)</span>
<span class="token variable">neg</span> <span class="token operator">&lt;-</span> <span class="token variable">neg</span><span class="token punctuation">[</span><span class="token function">order</span><span class="token punctuation">(</span><span class="token variable">neg</span><span class="token punctuation">$</span><span class="token property">start</span><span class="token punctuation">,</span> <span class="token variable">decreasing</span> <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span> 
<span class="token variable">neg</span> <span class="token operator">&lt;-</span> <span class="token variable">neg</span><span class="token punctuation">[</span><span class="token operator">!</span><span class="token function">duplicated</span><span class="token punctuation">(</span><span class="token variable">neg</span><span class="token punctuation">$</span><span class="token property">transcript</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span> <span class="token comment" spellcheck="true"># remove all but the first exons per transcript</span>
<span class="token variable">neg</span><span class="token punctuation">$</span><span class="token property">start</span> <span class="token operator">&lt;-</span> <span class="token variable">neg</span><span class="token punctuation">$</span><span class="token property">end</span> <span class="token operator">-</span> <span class="token number">1</span>

<span class="token variable">gene_annotation_sub</span> <span class="token operator">&lt;-</span> <span class="token function">rbind</span><span class="token punctuation">(</span><span class="token variable">pos</span><span class="token punctuation">,</span> <span class="token variable">neg</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># Make a subset of the TSS annotation columns containing just the coordinates </span>
<span class="token comment" spellcheck="true"># and the gene name</span>
<span class="token variable">gene_annotation_sub</span> <span class="token operator">&lt;-</span> <span class="token variable">gene_annotation_sub</span><span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token function">c</span><span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">:</span><span class="token number">3</span><span class="token punctuation">,</span> <span class="token number">8</span><span class="token punctuation">)</span><span class="token punctuation">]</span>

<span class="token comment" spellcheck="true"># Rename the gene symbol column to "gene"</span>
<span class="token function">names</span><span class="token punctuation">(</span><span class="token variable">gene_annotation_sub</span><span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">4</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"gene"</span>

<span class="token variable">input_cds</span> <span class="token operator">&lt;-</span> <span class="token function">annotate_cds_by_site</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">,</span> <span class="token variable">gene_annotation_sub</span><span class="token punctuation">)</span>

<span class="token function">head</span><span class="token punctuation">(</span><span class="token function">fData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true">#                               site_name chr    bp1    bp2 num_cells_expressed overlap          gene</span>
<span class="token comment" spellcheck="true"># chr18_10025_10225     chr18_10025_10225  18  10025  10225                   5      NA          &lt;NA&gt;</span>
<span class="token comment" spellcheck="true"># chr18_10603_11103     chr18_10603_11103  18  10603  11103                   1       1    AP005530.1</span>
<span class="token comment" spellcheck="true"># chr18_11604_13986     chr18_11604_13986  18  11604  13986                   9     203    AP005530.1</span>
<span class="token comment" spellcheck="true"># chr18_49557_50057     chr18_49557_50057  18  49557  50057                   2     331 RP11-683L23.1</span>
<span class="token comment" spellcheck="true"># chr18_50240_50740     chr18_50240_50740  18  50240  50740                   2     129 RP11-683L23.1</span>
<span class="token comment" spellcheck="true"># chr18_104385_104585 chr18_104385_104585  18 104385 104585                   1      NA          &lt;NA&gt;</span>

<span class="token comment" spellcheck="true">#### Generate gene activity scores ####</span>
<span class="token comment" spellcheck="true"># generate unnormalized gene activity matrix</span>
<span class="token variable">unnorm_ga</span> <span class="token operator">&lt;-</span> <span class="token function">build_gene_activity_matrix</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">,</span> <span class="token variable">conns</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># remove any rows/columns with all zeroes</span>
<span class="token variable">unnorm_ga</span> <span class="token operator">&lt;-</span> <span class="token variable">unnorm_ga</span><span class="token punctuation">[</span><span class="token operator">!</span><span class="token namespace">Matrix</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">rowSums</span><span class="token punctuation">(</span><span class="token variable">unnorm_ga</span><span class="token punctuation">)</span> <span class="token operator">==</span> <span class="token number">0</span><span class="token punctuation">,</span> <span class="token operator">!</span><span class="token namespace">Matrix</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">colSums</span><span class="token punctuation">(</span><span class="token variable">unnorm_ga</span><span class="token punctuation">)</span> <span class="token operator">==</span> <span class="token number">0</span><span class="token punctuation">]</span>

<span class="token comment" spellcheck="true"># make a list of num_genes_expressed</span>
<span class="token variable">num_genes</span> <span class="token operator">&lt;-</span> <span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">$</span><span class="token property">num_genes_expressed</span>
<span class="token function">names</span><span class="token punctuation">(</span><span class="token variable">num_genes</span><span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token function">row.names</span><span class="token punctuation">(</span><span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># normalize</span>
<span class="token variable">cicero_gene_activities</span> <span class="token operator">&lt;-</span> <span class="token function">normalize_gene_activities</span><span class="token punctuation">(</span><span class="token variable">unnorm_ga</span><span class="token punctuation">,</span> <span class="token variable">num_genes</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># if you had two datasets to normalize, you would pass both:</span>
<span class="token comment" spellcheck="true"># num_genes should then include all cells from both sets</span>
<span class="token variable">unnorm_ga2</span> <span class="token operator">&lt;-</span> <span class="token variable">unnorm_ga</span>
<span class="token variable">cicero_gene_activities</span> <span class="token operator">&lt;-</span> <span class="token function">normalize_gene_activities</span><span class="token punctuation">(</span><span class="token function">list</span><span class="token punctuation">(</span><span class="token variable">unnorm_ga</span><span class="token punctuation">,</span> <span class="token variable">unnorm_ga2</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token variable">num_genes</span><span class="token punctuation">)</span>


      </code></pre></figure>


      <h3 id="advanced-visualizaton">Advanced visualizaton</h3>

      <h4>Some useful parameters</h4>
      <p>
        Here, we will show some example of plotting parameters you may want to use to further explore your data 
        visually. Find documentation of all of the parameters in R's manual pages using 
        <code>?plot_connections</code>.
      </p>

      <p>
        <strong>Viewpoints:</strong> the <code>viewpoint</code> parameter lets you view only the connections 
        originating from a specific place in the genome. This might be useful in comparing data with 4C-seq data.
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">plot_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token number">10000</span><span class="token punctuation">,</span> <span class="token number">112367</span><span class="token punctuation">,</span> 
                <span class="token variable">viewpoint</span> <span class="token operator">=</span> <span class="token string">"chr18_48000_53000"</span><span class="token punctuation">,</span>
                <span class="token variable">gene_model</span> <span class="token operator">=</span> <span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> 
                <span class="token variable">coaccess_cutoff</span> <span class="token operator">=</span> <span class="token number">0</span><span class="token punctuation">,</span>
                <span class="token variable">connection_width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_track</span> <span class="token operator">=</span> <span class="token variable">chia_conns</span><span class="token punctuation">,</span>
                <span class="token variable">include_axis_track</span> <span class="token operator">=</span> <span class="token boolean">F</span><span class="token punctuation">,</span>
                <span class="token variable">collapseTranscripts</span> <span class="token operator">=</span> <span class="token string">"longest"</span><span class="token punctuation">)</span> 
      </code></pre></figure>

      <div class="text-center">
          <img src="./Cicero_files/viewpoint_example.png">
      </div>
      <p>
          <strong><code>alpha_by_coaccess</code>:</strong> the <code>alpha_by_coaccess</code> parameter lets you is 
          useful when you are concerned about overplotting. This parameter causes the alpha (transparency) of the 
          connection curves to be scaled based on the magnitude of the co-accessibility. Compare the following two 
          plots.
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">plot_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> 
                <span class="token variable">alpha_by_coaccess</span> <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">,</span> 
                <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token number">8575097</span><span class="token punctuation">,</span> <span class="token number">8839855</span><span class="token punctuation">,</span> 
                <span class="token variable">gene_model</span> <span class="token operator">=</span> <span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> 
                <span class="token variable">coaccess_cutoff</span> <span class="token operator">=</span> <span class="token number">0.1</span><span class="token punctuation">,</span> 
                <span class="token variable">connection_width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span> 
                <span class="token variable">collapseTranscripts</span> <span class="token operator">=</span> <span class="token string">"longest"</span> <span class="token punctuation">)</span>

<span class="token function">plot_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> 
                <span class="token variable">alpha_by_coaccess</span> <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span> 
                <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token number">8575097</span><span class="token punctuation">,</span> <span class="token number">8839855</span><span class="token punctuation">,</span> 
                <span class="token variable">gene_model</span> <span class="token operator">=</span> <span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> 
                <span class="token variable">coaccess_cutoff</span> <span class="token operator">=</span> <span class="token number">0.1</span><span class="token punctuation">,</span> 
                <span class="token variable">connection_width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span> 
                <span class="token variable">collapseTranscripts</span> <span class="token operator">=</span> <span class="token string">"longest"</span> <span class="token punctuation">)</span>
      </code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/alpha_by_co_F_example.png">
      </div>
      <div class="text-center">
        <img src="./Cicero_files/alpha_by_co_example.png">
      </div>    


      <p>
        <strong>Colors:</strong> There are several parameters having to do with colors:
        </p><ul>
          <li><code>peak_color</code></li>
          <li><code>comparison_peak_color</code></li>
          <li><code>connection_color</code></li>
          <li><code>comparison_connection_color</code></li>
          <li><code>gene_model_color</code></li>
          <li><code>viewpoint_color</code></li>
          <li><code>viewpoint_fill</code></li>
        </ul>
        Each of these parameters can be given color values in the form of color names, ex: "blue", or color codes, 
        ex: "#B4656F". Alternatively, for the first four parameters above, you can provide the name of a column in 
        your conns table that should be used to determine colors. In the case of <code>peak_color</code> and 
        <code>comparison_peak_color</code>, the column should correspond to <strong>Peak1</strong> color assignment.
      <p></p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token comment" spellcheck="true"># When the color column is not already colors, random colors are assigned</span>
<span class="token function">plot_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> 
                <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token number">10000</span><span class="token punctuation">,</span> <span class="token number">112367</span><span class="token punctuation">,</span>
                <span class="token variable">connection_color</span> <span class="token operator">=</span> <span class="token string">"in_chia_100"</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_track</span> <span class="token operator">=</span> <span class="token variable">chia_conns</span><span class="token punctuation">,</span>
                <span class="token variable">peak_color</span> <span class="token operator">=</span> <span class="token string">"green"</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_peak_color</span> <span class="token operator">=</span> <span class="token string">"orange"</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_connection_color</span> <span class="token operator">=</span> <span class="token string">"purple"</span><span class="token punctuation">,</span>
                <span class="token variable">gene_model_color</span> <span class="token operator">=</span> "<span class="token comment" spellcheck="true">#2DD881",</span>
                <span class="token variable">gene_model</span> <span class="token operator">=</span> <span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> 
                <span class="token variable">coaccess_cutoff</span> <span class="token operator">=</span> <span class="token number">0.1</span><span class="token punctuation">,</span> 
                <span class="token variable">connection_width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span> 
                <span class="token variable">collapseTranscripts</span> <span class="token operator">=</span> <span class="token string">"longest"</span> <span class="token punctuation">)</span>
      </code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/colors1_example.png">
      </div>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token comment" spellcheck="true"># If I want specific color scheme, I can make a column of color names</span>
<span class="token variable">conns</span><span class="token punctuation">$</span><span class="token property">conn_color</span> <span class="token operator">&lt;-</span> <span class="token string">"orange"</span>
<span class="token variable">conns</span><span class="token punctuation">$</span><span class="token property">conn_color</span><span class="token punctuation">[</span><span class="token variable">conns</span><span class="token punctuation">$</span><span class="token property">in_chia_100</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"green"</span>
<span class="token function">plot_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> 
                <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token number">10000</span><span class="token punctuation">,</span> <span class="token number">112367</span><span class="token punctuation">,</span>
                <span class="token variable">connection_color</span> <span class="token operator">=</span> <span class="token string">"conn_color"</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_track</span> <span class="token operator">=</span> <span class="token variable">chia_conns</span><span class="token punctuation">,</span>
                <span class="token variable">peak_color</span> <span class="token operator">=</span> <span class="token string">"green"</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_peak_color</span> <span class="token operator">=</span> <span class="token string">"orange"</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_connection_color</span> <span class="token operator">=</span> <span class="token string">"purple"</span><span class="token punctuation">,</span>
                <span class="token variable">gene_model_color</span> <span class="token operator">=</span> "<span class="token comment" spellcheck="true">#2DD881",</span>
                <span class="token variable">gene_model</span> <span class="token operator">=</span> <span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> 
                <span class="token variable">coaccess_cutoff</span> <span class="token operator">=</span> <span class="token number">0.1</span><span class="token punctuation">,</span> 
                <span class="token variable">connection_width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span> 
                <span class="token variable">collapseTranscripts</span> <span class="token operator">=</span> <span class="token string">"longest"</span> <span class="token punctuation">)</span>
      </code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/colors2_example.png">
      </div>
      
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token comment" spellcheck="true"># For coloring Peaks, I need the color column to correspond to Peak1:</span>
<span class="token variable">conns</span><span class="token punctuation">$</span><span class="token property">peak1_color</span> <span class="token operator">&lt;-</span> <span class="token boolean">FALSE</span>
<span class="token variable">conns</span><span class="token punctuation">$</span><span class="token property">peak1_color</span><span class="token punctuation">[</span><span class="token variable">conns</span><span class="token punctuation">$</span><span class="token property">Peak1</span> <span class="token operator">==</span> <span class="token string">"chr18_11604_13986"</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token boolean">TRUE</span>
<span class="token function">plot_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> 
                <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token number">10000</span><span class="token punctuation">,</span> <span class="token number">112367</span><span class="token punctuation">,</span>
                <span class="token variable">connection_color</span> <span class="token operator">=</span> <span class="token string">"green"</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_track</span> <span class="token operator">=</span> <span class="token variable">chia_conns</span><span class="token punctuation">,</span>
                <span class="token variable">peak_color</span> <span class="token operator">=</span> <span class="token string">"peak1_color"</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_peak_color</span> <span class="token operator">=</span> <span class="token string">"orange"</span><span class="token punctuation">,</span>
                <span class="token variable">comparison_connection_color</span> <span class="token operator">=</span> <span class="token string">"purple"</span><span class="token punctuation">,</span>
                <span class="token variable">gene_model_color</span> <span class="token operator">=</span> "<span class="token comment" spellcheck="true">#2DD881",</span>
                <span class="token variable">gene_model</span> <span class="token operator">=</span> <span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> 
                <span class="token variable">coaccess_cutoff</span> <span class="token operator">=</span> <span class="token number">0.1</span><span class="token punctuation">,</span> 
                <span class="token variable">connection_width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span> 
                <span class="token variable">collapseTranscripts</span> <span class="token operator">=</span> <span class="token string">"longest"</span> <span class="token punctuation">)</span>

      </code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/colors3_example.png">
      </div>
      

      <h4>Customizing everything with <code>return_as_list</code></h4>
      <p>
        We realize that you may wish to plot other kinds of data along with Cicero connections. Because of this, we 
        include the <code>return_as_list</code> option. This parameter allows you to use the extensive functionality 
        of <a href="https://bioconductor.org/packages/release/bioc/html/Gviz.html">Gviz</a> to plot many different 
        kinds of genomic data. See the 
        <a href="https://bioconductor.org/packages/release/bioc/vignettes/Gviz/inst/doc/Gviz.html">Gviz user guide</a> 
        to see the many options.
      </p>
      <p>
        If <code>return_as_list</code></p> is set to <code>TRUE</code>, an image will not be plotted, and instead, a 
        list of plot components will be returned. Let's go back to a simpler example:
      <p></p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">plot_list</span> <span class="token operator">&lt;-</span> <span class="token function">plot_connections</span><span class="token punctuation">(</span><span class="token variable">conns</span><span class="token punctuation">,</span> 
                    <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token number">10000</span><span class="token punctuation">,</span> <span class="token number">112367</span><span class="token punctuation">,</span>
                    <span class="token variable">gene_model</span> <span class="token operator">=</span> <span class="token variable">gene_annotation_sample</span><span class="token punctuation">,</span> 
                    <span class="token variable">coaccess_cutoff</span> <span class="token operator">=</span> <span class="token number">0.1</span><span class="token punctuation">,</span> 
                    <span class="token variable">connection_width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span> 
                    <span class="token variable">collapseTranscripts</span> <span class="token operator">=</span> <span class="token string">"longest"</span><span class="token punctuation">,</span> 
                    <span class="token variable">return_as_list</span> <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
<span class="token variable">plot_list</span>
<span class="token comment" spellcheck="true"># [[1]]</span>
<span class="token comment" spellcheck="true"># CustomTrack ''</span>

<span class="token comment" spellcheck="true"># [[2]]</span>
<span class="token comment" spellcheck="true"># AnnotationTrack 'Peaks'</span>
<span class="token comment" spellcheck="true"># | genome: NA</span>
<span class="token comment" spellcheck="true"># | active chromosome: chr18</span>
<span class="token comment" spellcheck="true"># | annotation features: 7</span>

<span class="token comment" spellcheck="true"># [[3]]</span>
<span class="token comment" spellcheck="true"># Genome axis 'Axis'</span>

<span class="token comment" spellcheck="true"># [[4]]</span>
<span class="token comment" spellcheck="true"># GeneRegionTrack 'Gene Model'</span>
<span class="token comment" spellcheck="true"># | genome: NA</span>
<span class="token comment" spellcheck="true"># | active chromosome: chr18</span>
<span class="token comment" spellcheck="true"># | annotation features: 33</span>
      </code></pre></figure>
      <p>
        This made a list of components in the order of plotting. First is the CustomTrack, which is the Cicero 
        connections, next is the peak annotation track, next is the genome axis track, and lastly is the gene model 
        track. I can now add another track, rearrange tracks, or replace tracks using Gviz:
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">conservation</span> <span class="token operator">&lt;-</span> <span class="token function">UcscTrack</span><span class="token punctuation">(</span><span class="token variable">genome</span> <span class="token operator">=</span> <span class="token string">"hg19"</span><span class="token punctuation">,</span> <span class="token variable">chromosome</span> <span class="token operator">=</span> <span class="token string">"chr18"</span><span class="token punctuation">,</span>
                        <span class="token variable">track</span> <span class="token operator">=</span> <span class="token string">"Conservation"</span><span class="token punctuation">,</span> <span class="token variable">table</span> <span class="token operator">=</span> <span class="token string">"phyloP100wayAll"</span><span class="token punctuation">,</span>
                        <span class="token variable">fontsize</span><span class="token punctuation">.</span><span class="token variable">group</span><span class="token operator">=</span><span class="token number">6</span><span class="token punctuation">,</span><span class="token variable">fontsize</span><span class="token operator">=</span><span class="token number">6</span><span class="token punctuation">,</span> <span class="token variable">cex</span><span class="token punctuation">.</span><span class="token variable">axis</span><span class="token operator">=</span><span class="token punctuation">.</span><span class="token number">8</span><span class="token punctuation">,</span>
                        <span class="token variable">from</span> <span class="token operator">=</span> <span class="token number">10000</span><span class="token punctuation">,</span> <span class="token variable">to</span> <span class="token operator">=</span> <span class="token number">112367</span><span class="token punctuation">,</span> <span class="token variable">trackType</span> <span class="token operator">=</span> <span class="token string">"DataTrack"</span><span class="token punctuation">,</span>
                        <span class="token variable">start</span> <span class="token operator">=</span> <span class="token string">"start"</span><span class="token punctuation">,</span> <span class="token variable">end</span> <span class="token operator">=</span> <span class="token string">"end"</span><span class="token punctuation">,</span> <span class="token variable">data</span> <span class="token operator">=</span> <span class="token string">"score"</span><span class="token punctuation">,</span> <span class="token variable">size</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">1</span><span class="token punctuation">,</span>
                        <span class="token variable">type</span> <span class="token operator">=</span> <span class="token string">"histogram"</span><span class="token punctuation">,</span> <span class="token variable">window</span> <span class="token operator">=</span> <span class="token string">"auto"</span><span class="token punctuation">,</span> <span class="token variable">col</span><span class="token punctuation">.</span><span class="token variable">histogram</span> <span class="token operator">=</span> "<span class="token comment" spellcheck="true">#587B7F",</span>
                        <span class="token variable">fill</span><span class="token punctuation">.</span><span class="token variable">histogram</span> <span class="token operator">=</span> "<span class="token comment" spellcheck="true">#587B7F", ylim = c(-1, 2.5),</span>
                        <span class="token variable">name</span> <span class="token operator">=</span> <span class="token string">"Conservation"</span><span class="token punctuation">)</span>
<span class="token comment" spellcheck="true"># I will replace the genome axis track with a track on conservation values</span>
<span class="token variable">plot_list</span><span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token variable">conservation</span>   

<span class="token comment" spellcheck="true"># To make the plot, I will now use Gviz's plotTracks function</span>
<span class="token comment" spellcheck="true"># The included options are the defaults in plot_connections, </span>
<span class="token comment" spellcheck="true"># but all can be modified according to Gviz's documentation</span>

<span class="token comment" spellcheck="true"># The main new paramter that you must include, is the sizes</span>
<span class="token comment" spellcheck="true"># parameter. This parameter controls what proportion of the</span>
<span class="token comment" spellcheck="true"># height of your plot is allocated for each track. The sizes</span>
<span class="token comment" spellcheck="true"># parameter must be a vector of the same length as plot_list</span>

<span class="token namespace">Gviz</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">plotTracks</span><span class="token punctuation">(</span><span class="token variable">plot_list</span><span class="token punctuation">,</span>  
                <span class="token variable">sizes</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token number">2</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
                <span class="token variable">from</span> <span class="token operator">=</span> <span class="token number">10000</span><span class="token punctuation">,</span> <span class="token variable">to</span> <span class="token operator">=</span> <span class="token number">112367</span><span class="token punctuation">,</span> <span class="token variable">chromosome</span> <span class="token operator">=</span> <span class="token string">"chr18"</span><span class="token punctuation">,</span> 
                <span class="token variable">transcriptAnnotation</span> <span class="token operator">=</span> <span class="token string">"symbol"</span><span class="token punctuation">,</span>
                <span class="token variable">col</span><span class="token punctuation">.</span><span class="token variable">axis</span> <span class="token operator">=</span> <span class="token string">"black"</span><span class="token punctuation">,</span> 
                <span class="token variable">fontsize</span><span class="token punctuation">.</span><span class="token variable">group</span> <span class="token operator">=</span> <span class="token number">6</span><span class="token punctuation">,</span>
                <span class="token variable">fontcolor</span><span class="token punctuation">.</span><span class="token variable">legend</span> <span class="token operator">=</span> <span class="token string">"black"</span><span class="token punctuation">,</span>
                <span class="token variable">lwd</span><span class="token operator">=</span><span class="token punctuation">.</span><span class="token number">3</span><span class="token punctuation">,</span>
                <span class="token variable">title</span><span class="token punctuation">.</span><span class="token variable">width</span> <span class="token operator">=</span> <span class="token punctuation">.</span><span class="token number">5</span><span class="token punctuation">,</span>
                <span class="token variable">background</span><span class="token punctuation">.</span><span class="token variable">title</span> <span class="token operator">=</span> <span class="token string">"transparent"</span><span class="token punctuation">,</span> 
                <span class="token variable">col</span><span class="token punctuation">.</span><span class="token variable">border</span><span class="token punctuation">.</span><span class="token variable">title</span> <span class="token operator">=</span> <span class="token string">"transparent"</span><span class="token punctuation">)</span>
      </code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/return_as_list_example.png">
      </div>


      

      <h3 id="important-considerations-for-non-human-data">Important Considerations for Non-Human Data</h3>

      <p>
        Cicero's default parameters were designed for use in human and mouse. There are a few parameters that we 
        expect will need to be changed for use in different model organisms. Here are the parameters along with a 
        brief discussion of each:

      </p><div class="panel panel-info">
          <div class="panel-heading">
            <h4 class="panel-title">Parameters to consider for non-human data</h4>
          </div>
          <div class="panel-body">
            <table class="table">
              <thead>
                <tr>
                  <th>Parameter</th>
                  <th>Functions where used</th>
                  <th>Notes</th>
                  <th>Potential values</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td><code>s</code></td>
                  <td><code>estimate_distance_parameter</code>, <code>generate_cicero_models</code></td>
                  <td>As discussed previously, the value <code>s</code> is the scaling exponent of the power-law 
                    function that estimates the global decay in contact frequency. This value is generally estimated 
                    using Hi-C contact frequencies. As a default, we have used 0.75, which corresponds to the 
                    tension globule polymer model of DNA (Sanborn et al., 2015). However, in other organisms, 
                    different decay functions have been observed. For example, in Drosophila, a scaling value of 0.85 
                    has been proposed (Sexton et al., 2012).
                  </td>
                  <td>Human and mouse: 0.75<br>Drosophila: 0.85</td>
                </tr>
                <tr>
                  <td><code>distance_constraint</code></td>
                  <td><code>estimate_distance_parameter</code></td>
                  <td>The value <code>distance_constraint</code> is the distance in base pairs at beyond which few
                    (less than 5%) meaningful <i>cis</i>-regulatory connections are expected to exist. While there are 
                    examples of very long range <i>cis</i>-regulatory contacts in various organisms, this value should be 
                    more of a general rule as Cicero can still detect longer range connections with high 
                    co-accessibility.
                  </td>
                  <td>Human and mouse: 250,000<br>Drosophila: 50,000</td>
                  </tr>
                  <tr>
                    <td><code>window</code></td>
                    <td><code>estimate_distance_parameter</code>, <code>generate_cicero_models</code></td>
                    <td>The <code>window</code> value is the length in base pairs of the sliding window that Cicero 
                      will use when measuring co-accessibility. This value must be significantly longer (&gt;30%) than 
                      the value for <code>distance_constraint</code>. This is the maximum distance of sites that 
                      will be compared.
                    </td>
                    <td>Human and mouse: 500,000<br>Drosophila: 100,000</td>
                  </tr>
              </tbody>
            </table>
          </div>
        </div>
      <p></p>

      <h2 id="single-cell-accessibility-trajectories">Single-cell accessibility trajectories</h2>
      <p>
        The second major function of the Cicero package is to extend Monocle 2 for use with single-cell accessibility 
        data. The main obstacle to overcome with chromatin accessibility data is the sparsity, so most of the 
        extensions and methods are designed to address that.
      </p>

      <h3 id="constructing-trajectories-with-accessibility-data">Constructing trajectories with accessibility data</h3>

      <p>We strongly recommend that you consult the Monocle website, especially 
        <a href="http://cole-trapnell-lab.github.io/monocle-release/docs/#constructing-single-cell-trajectories">
        this section</a> prior to reading about Cicero's extension of the Monocle analysis described. Briefly, 
        Monocle infers pseudotime trajectories in three steps:
        </p><ol>
          <li>Choosing sites that define progress</li>
          <li>Reducing the dimensionality of the data</li>
          <li>Ordering cells in pseudotime</li>
        </ol>
        We will describe how each piece is modified for use with sparse single-cell chromatin accessibility data.
      <p></p>

      <h4>Aggregation: the primary method for addressing sparsity</h4>
      <p>
        The primary way that the Cicero package deals with the sparsity of single-cell chromatin accessibility data 
        is through aggregation. Aggregating the counts of either single cells or single peaks allows us to produce a 
        "consensus" count matrix, reducing noise and allowing us to move out of the binary regime. Under this 
        grouping, the number of cells in which a particular site is accessible can be modeled with a binomial 
        distribution or, for sufficiently large groups, the corresponding Gaussian approximation. Modeling grouped 
        accessibility counts as normally distributed allows Cicero to easily adjust them for arbitrary technical 
        covariates by simply fitting a linear model and taking the residuals with respect to it as the adjusted 
        accessibility score for each group of cells. We demonstrate how to apply this grouping practically below.
      </p>

      <h5><code>aggregate_nearby_peaks</code></h5>
      <p>
        The primary aggregation used for trajectory reconstruction is betweeen nearby peaks. This keeps single cells 
        separate while aggregating regions of the genome and looking for chromatin accessibility within them. The 
        function <code>aggregate_nearby_peaks</code> finds sites within a certain distance of each other and 
        aggregates them together by summing their counts. Depending on the density of your data, you may want to try 
        different <code>distance</code> parameters. In published work we have used 1000 and 10,000.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">data</span><span class="token punctuation">(</span><span class="token string">"cicero_data"</span><span class="token punctuation">)</span>
<span class="token variable">input_cds</span> <span class="token operator">&lt;-</span> <span class="token function">make_atac_cds</span><span class="token punctuation">(</span><span class="token variable">cicero_data</span><span class="token punctuation">,</span> <span class="token variable">binarize</span> <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># Add some cell meta-data</span>
<span class="token function">data</span><span class="token punctuation">(</span><span class="token string">"cell_data"</span><span class="token punctuation">)</span>
<span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token function">cbind</span><span class="token punctuation">(</span><span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token variable">cell_data</span><span class="token punctuation">[</span><span class="token function">row.names</span><span class="token punctuation">(</span><span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span><span class="token punctuation">)</span>
<span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">$</span><span class="token property">cell</span> <span class="token operator">&lt;-</span> <span class="token boolean">NULL</span>

<span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">aggregate_nearby_peaks</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">,</span> <span class="token variable">distance</span> <span class="token operator">=</span> <span class="token number">10000</span><span class="token punctuation">)</span>
<span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">detectGenes</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">)</span>
<span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">estimateSizeFactors</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">)</span>
<span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">estimateDispersions</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">)</span></code></pre></figure>

      <h4>Choosing sites that define progress</h4>

      <p>
        There are several options for choosing the sites to use during dimensionality reduction. Monocle has a 
        discussion about the options 
        <a href="http://cole-trapnell-lab.github.io/monocle-release/docs/#alternative-choices-for-ordering-genes">
          here</a>. Any of these options could be used with your new aggregated CDS, depending on the information you 
        have a available in your dataset. Here, we will show two examples:
      </p>

      <h5 data-toc-text="Choose sites by differential analysis">Choose sites by differential analysis<span class="label label-success">Alternative</span></h5>

      <p>
        If your data has defined beginning and end points, you can determine which sites define progress by a 
        differential accessibility test. We use Monocle's <code>differentialGeneTest</code> function looking for 
        sites that are different in the timepoint groups.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token comment" spellcheck="true"># This takes a few minutes to run</span>
<span class="token variable">diff_timepoint</span> <span class="token operator">&lt;-</span> <span class="token function">differentialGeneTest</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span>
                    <span class="token variable">fullModelFormulaStr</span><span class="token operator">=</span><span class="token string">"~timepoint + num_genes_expressed"</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># We chose a very high q-value cutoff because there are</span>
<span class="token comment" spellcheck="true"># so few sites in the sample dataset, in general a q-value</span>
<span class="token comment" spellcheck="true"># cutoff in the range of 0.01 to 0.1 would be appropriate</span>
<span class="token variable">ordering_sites</span> <span class="token operator">&lt;-</span> <span class="token function">row.names</span><span class="token punctuation">(</span><span class="token function">subset</span><span class="token punctuation">(</span><span class="token variable">diff_timepoint</span><span class="token punctuation">,</span> <span class="token variable">qval</span> <span class="token operator">&lt;</span> <span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">)</span></code></pre></figure>

      <h5 data-toc-text="Choose sites by dpFeature">Choose sites by dpFeature <span class="label label-success">Alternative</span></h5>
      <p>
        Alternatively, you can choose sites for dimensionality reduction by using Monocle's 
        <a href="http://cole-trapnell-lab.github.io/monocle-release/docs/#dpfeature-selecting-features-from-dense-cell-clusters">
        dpFeature</a> method. dpFeature chooses sites based on how they differ among clusters of cells. Here, we give 
        some example code reproduced from Monocle, for more information, see the 
        <a href="http://cole-trapnell-lab.github.io/monocle-release/docs/#alternative-choices-for-ordering-genes">
          Monocle description</a>.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">plot_pc_variance_explained</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token variable">return_all</span> <span class="token operator">=</span> <span class="token boolean">F</span><span class="token punctuation">)</span> <span class="token comment" spellcheck="true">#Choose 2 PCs</span></code></pre></figure>
      <div class="text-center">
        <img src="./Cicero_files/plot_pc_var.png">
      </div>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">reduceDimension</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span>
                            <span class="token variable">max_components</span> <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">,</span>
                            <span class="token variable">norm_method</span> <span class="token operator">=</span> <span class="token string">'log'</span><span class="token punctuation">,</span>
                            <span class="token variable">num_dim</span> <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">,</span>
                            <span class="token variable">reduction_method</span> <span class="token operator">=</span> <span class="token string">'tSNE'</span><span class="token punctuation">,</span>
                            <span class="token variable">verbose</span> <span class="token operator">=</span> <span class="token boolean">T</span><span class="token punctuation">)</span>

<span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">clusterCells</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token variable">verbose</span> <span class="token operator">=</span> <span class="token boolean">F</span><span class="token punctuation">)</span>

<span class="token function">plot_cell_clusters</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token variable">color_by</span> <span class="token operator">=</span> <span class="token string">'as.factor(Cluster)'</span><span class="token punctuation">)</span></code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/plot_cell_clust.png">
      </div>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">clustering_DA_sites</span> <span class="token operator">&lt;-</span> <span class="token function">differentialGeneTest</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token comment" spellcheck="true">#Takes a few minutes</span>
                                          <span class="token variable">fullModelFormulaStr</span> <span class="token operator">=</span> <span class="token string">'~Cluster'</span><span class="token punctuation">)</span>

<span class="token variable">ordering_sites</span> <span class="token operator">&lt;-</span>
<span class="token function">row.names</span><span class="token punctuation">(</span><span class="token variable">clustering_DA_sites</span><span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token function">order</span><span class="token punctuation">(</span><span class="token variable">clustering_DA_sites</span><span class="token punctuation">$</span><span class="token property">qval</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">:</span><span class="token number">1000</span><span class="token punctuation">]</span></code></pre></figure>

      <h4>Reduce the dimensionality of the data and order cells</h4>
      <p>
        However you chose your ordering sites, the first step of dimensionality reduction is to use 
        <code>setOrderingFilter</code> to mark the sites you want to use for dimesionality reduction. In the 
        following figures, we are using the ordering_sites from 
        <strong>Choose sites by differential analysis</strong> above.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">setOrderingFilter</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token variable">ordering_sites</span><span class="token punctuation">)</span></code></pre></figure>

      <p>
        Next, we use DDRTree to reduce dimensionality and then order the cells along the trajectory. Importantly, we 
        use num_genes_expressed in our residual model formula to account for overall assay efficiency.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">reduceDimension</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token variable">max_components</span> <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">,</span>
        <span class="token variable">residualModelFormulaStr</span><span class="token operator">=</span><span class="token string">"~as.numeric(num_genes_expressed)"</span><span class="token punctuation">,</span>
        <span class="token variable">reduction_method</span> <span class="token operator">=</span> <span class="token string">'DDRTree'</span><span class="token punctuation">)</span>
<span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">orderCells</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">)</span>
<span class="token function">plot_cell_trajectory</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token variable">color_by</span> <span class="token operator">=</span> <span class="token string">"timepoint"</span><span class="token punctuation">)</span></code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/HSMM_trajectory_example.png">
      </div>

      <p>
        Once you have a cell trajectory, you need to make sure that pseudotime proceeds how you expect. In our 
        example, we want pseudotime to start where most of the time 0 cells are located and proceed towards the later 
        timepoints. Further information on this can be found 
        <a href="http://cole-trapnell-lab.github.io/monocle-release/docs/#constructing-single-cell-trajectories">
          here</a> on the Monocle website. We first color our trajectory plot by State, which is how Monocle assigns 
        segments of the tree.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">plot_cell_trajectory</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token variable">color_by</span> <span class="token operator">=</span> <span class="token string">"State"</span><span class="token punctuation">)</span></code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/HSMM_trajectory_example_state.png">
      </div>

      <p>
        From this plot, we can see that the beginning of pseudotime should be from state 4. We now reorder cells 
        setting the root state to 4. We can then check that the ordering makes sense by coloring the plot by 
        Pseudotime.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">agg_cds</span> <span class="token operator">&lt;-</span> <span class="token function">orderCells</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token variable">root_state</span> <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">)</span>
<span class="token function">plot_cell_trajectory</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">,</span> <span class="token variable">color_by</span> <span class="token operator">=</span> <span class="token string">"Pseudotime"</span><span class="token punctuation">)</span></code></pre></figure>
      <div class="text-center">
        <img src="./Cicero_files/HSMM_trajectory_example_pseudo.png">
      </div>

      <p>
        Now that we have Pseudotime values for each cell (<code>pData(agg_cds)$Pseudotime</code>), we need to assign 
        these values back to our original CDS object. In addition, we will assign the State information back to the 
        original CDS.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">$</span><span class="token property">Pseudotime</span> <span class="token operator">&lt;-</span> <span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token function">colnames</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span><span class="token punctuation">$</span><span class="token property">Pseudotime</span>
<span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">$</span><span class="token property">State</span> <span class="token operator">&lt;-</span> <span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">agg_cds</span><span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token function">colnames</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span><span class="token punctuation">$</span><span class="token property">State</span></code></pre></figure>

      <h3 id="differential-accessibility-analysis">Differential Accessibility Analysis</h3>
      <p>
        Once you have your cells ordered in pseudotime, you can ask where in the genome chromatin accessibility is 
        changing across time. If you know of specific sites that are important to your system, you may want to 
        visualize the accessibility at those sites across pseudotime.
      </p>

      <h4>Visualizing accessibility across pseudotime</h4>
      <h5><code>plot_accessibility_in_pseudotime</code></h5>
      <p>
        For simplicity, we will exclude the branch in our trajectory to make our trajectory linear.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">input_cds_lin</span> <span class="token operator">&lt;-</span> <span class="token variable">input_cds</span><span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token function">row.names</span><span class="token punctuation">(</span><span class="token function">subset</span><span class="token punctuation">(</span><span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token variable">State</span>  <span class="token operator">!</span><span class="token operator">=</span> <span class="token number">5</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">]</span>

<span class="token function">plot_accessibility_in_pseudotime</span><span class="token punctuation">(</span><span class="token variable">input_cds_lin</span><span class="token punctuation">[</span><span class="token function">c</span><span class="token punctuation">(</span><span class="token string">"chr18_38156577_38158261"</span><span class="token punctuation">,</span> <span class="token string">"chr18_48373358_48374180"</span><span class="token punctuation">,</span> <span class="token string">"chr18_60457956_60459080"</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">)</span></code></pre></figure>

      <div class="text-center">
        <img src="./Cicero_files/example_plot_accessibility.png" width="400px">
      </div>

      <h4>Running <code>differentialGeneTest</code> with single cell chromatin accessibility data</h4>
      <p>
        In the previous section, we used aggregation of sites to discover cell-level information (cell pseudotime). 
        In this section, we are interested in a site-level statistic (whether a site is changing in pseudotime), so 
        we will aggregate similar cells. To do this, Cicero has a useful function called 
        <code>aggregate_by_cell_bin</code>.
      </p>

      <h5><code>aggregate_by_cell_bin</code></h5>
      <p>
        We use the function <code>aggregate_by_cell_bin</code> to aggregate our input CDS object by a column in the 
        pData table. In this example, we will assign cells to bins by cutting the pseudotime trajectory into 10 parts.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds_lin</span><span class="token punctuation">)</span><span class="token punctuation">$</span><span class="token property">cell_subtype</span> <span class="token operator">&lt;-</span> <span class="token function">cut</span><span class="token punctuation">(</span><span class="token function">pData</span><span class="token punctuation">(</span><span class="token variable">input_cds_lin</span><span class="token punctuation">)</span><span class="token punctuation">$</span><span class="token property">Pseudotime</span><span class="token punctuation">,</span> <span class="token number">10</span><span class="token punctuation">)</span>
<span class="token variable">binned_input_lin</span> <span class="token operator">&lt;-</span><span class="token function">aggregate_by_cell_bin</span><span class="token punctuation">(</span><span class="token variable">input_cds_lin</span><span class="token punctuation">,</span> <span class="token string">"cell_subtype"</span><span class="token punctuation">)</span></code></pre></figure>

      <p>
        We are now ready to run Monocle's differentialGeneTest function to find sites that are differentially 
        accessible across pseudotime. In this example, we include num_genes_expressed as a covariate to subtract its 
        effect.
      </p>

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">diff_test_res</span> <span class="token operator">&lt;-</span> <span class="token function">differentialGeneTest</span><span class="token punctuation">(</span><span class="token variable">binned_input_lin</span><span class="token punctuation">,</span>
  <span class="token variable">fullModelFormulaStr</span><span class="token operator">=</span><span class="token string">"~sm.ns(Pseudotime, df=3) + sm.ns(num_genes_expressed, df=3)"</span><span class="token punctuation">,</span>
  <span class="token variable">reducedModelFormulaStr</span><span class="token operator">=</span><span class="token string">"~sm.ns(num_genes_expressed, df=3)"</span><span class="token punctuation">,</span> <span class="token variable">cores</span><span class="token operator">=</span><span class="token number">1</span><span class="token punctuation">)</span></code></pre></figure>

      <h2 id="useful-functions">Useful Functions</h2>
      <h3 id="annotate-cds-by-site"><code>annotate_cds_by_site</code></h3>
      <p>
        It is often useful to add additional annotations about your peaks to your CDS object. For example, you might 
        want to know which peaks overlap an exon, or a transcription start site. Cicero includes the function 
        <code>annotate_cds_by_site</code>, which takes as input your CDS, and a data frame or file path with 
        bed-format information (chromosome, bp1, bp2, further columns). For example, suppose I have some data 
        (<code>feat</code>) about epigenetic marks in my system:
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">head</span><span class="token punctuation">(</span><span class="token function">fData</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true">#                               site_name chr    bp1    bp2 num_cells_expressed </span>
<span class="token comment" spellcheck="true"># chr18_10025_10225     chr18_10025_10225  18  10025  10225                   5 </span>
<span class="token comment" spellcheck="true"># chr18_10603_11103     chr18_10603_11103  18  10603  11103                   1 </span>
<span class="token comment" spellcheck="true"># chr18_11604_13986     chr18_11604_13986  18  11604  13986                   9  </span>
<span class="token comment" spellcheck="true"># chr18_49557_50057     chr18_49557_50057  18  49557  50057                   2   </span>
<span class="token comment" spellcheck="true"># chr18_50240_50740     chr18_50240_50740  18  50240  50740                   2   </span>
<span class="token comment" spellcheck="true"># chr18_104385_104585 chr18_104385_104585  18 104385 104585                   1  </span>

<span class="token variable">feat</span> <span class="token operator">&lt;-</span> <span class="token function">data.frame</span><span class="token punctuation">(</span><span class="token variable">chr</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token string">"chr18"</span><span class="token punctuation">,</span> <span class="token string">"chr18"</span><span class="token punctuation">)</span><span class="token punctuation">,</span> 
                  <span class="token variable">bp1</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token number">10000</span><span class="token punctuation">,</span> <span class="token number">10800</span><span class="token punctuation">,</span> <span class="token number">50000</span><span class="token punctuation">,</span> <span class="token number">100000</span><span class="token punctuation">)</span><span class="token punctuation">,</span> 
                  <span class="token variable">bp2</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token number">10700</span><span class="token punctuation">,</span> <span class="token number">11000</span><span class="token punctuation">,</span> <span class="token number">60000</span><span class="token punctuation">,</span> <span class="token number">110000</span><span class="token punctuation">)</span><span class="token punctuation">,</span> 
                  <span class="token variable">type</span> <span class="token operator">=</span> <span class="token function">c</span><span class="token punctuation">(</span><span class="token string">"Acetylated"</span><span class="token punctuation">,</span> <span class="token string">"Methylated"</span><span class="token punctuation">,</span> <span class="token string">"Acetylated"</span><span class="token punctuation">,</span> <span class="token string">"Methylated"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token function">head</span><span class="token punctuation">(</span><span class="token variable">feat</span><span class="token punctuation">)</span>
<span class="token comment" spellcheck="true">#     chr    bp1    bp2       type</span>
<span class="token comment" spellcheck="true"># 1 chr18  10000  10700 Acetylated</span>
<span class="token comment" spellcheck="true"># 2 chr18  10800  11000 Methylated</span>
<span class="token comment" spellcheck="true"># 3 chr18  50000  60000 Acetylated</span>
<span class="token comment" spellcheck="true"># 4 chr18 100000 110000 Methylated</span></code></pre></figure>
      <p>
        I can use annotate_cds_by_site to determine which of my peaks have those epigenetic marks.
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">temp</span> <span class="token operator">&lt;-</span> <span class="token function">annotate_cds_by_site</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">,</span> <span class="token variable">feat</span><span class="token punctuation">)</span>
<span class="token function">head</span><span class="token punctuation">(</span><span class="token function">fData</span><span class="token punctuation">(</span><span class="token variable">temp</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true">#                               site_name chr    bp1    bp2 num_cells_expressed overlap       type</span>
<span class="token comment" spellcheck="true"># chr18_10025_10225     chr18_10025_10225  18  10025  10225                   5     201 Acetylated</span>
<span class="token comment" spellcheck="true"># chr18_10603_11103     chr18_10603_11103  18  10603  11103                   1     201 Methylated</span>
<span class="token comment" spellcheck="true"># chr18_11604_13986     chr18_11604_13986  18  11604  13986                   9      NA       &lt;NA&gt;</span>
<span class="token comment" spellcheck="true"># chr18_49557_50057     chr18_49557_50057  18  49557  50057                   2      58 Acetylated</span>
<span class="token comment" spellcheck="true"># chr18_50240_50740     chr18_50240_50740  18  50240  50740                   2     501 Acetylated</span>
<span class="token comment" spellcheck="true"># chr18_104385_104585 chr18_104385_104585  18 104385 104585                   1     201 Methylated</span></code></pre></figure>

      <p>
        This function has added two columns compared to above. Overlap represents the number of base pairs 
        overlapping between the interval in <code>feat</code> and the given peak, and type is the assigned status 
        based on the overlap. If you look closely, site <code>chr18_10603_11103</code> was actually overlapped by two 
        intervals in <code>feat</code>. In its default mode, <code>annotate_cds_by_site</code> will choose the 
        largest overlapping interval to report. If you would like to see all overlapping intervals, you can set the 
        <code>all</code> flag to <code>TRUE</code>. 
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token variable">temp</span> <span class="token operator">&lt;-</span> <span class="token function">annotate_cds_by_site</span><span class="token punctuation">(</span><span class="token variable">input_cds</span><span class="token punctuation">,</span> <span class="token variable">feat</span><span class="token punctuation">,</span> <span class="token variable">all</span><span class="token operator">=</span><span class="token boolean">TRUE</span><span class="token punctuation">)</span>
<span class="token function">head</span><span class="token punctuation">(</span><span class="token function">fData</span><span class="token punctuation">(</span><span class="token variable">temp</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true">#                               site_name chr    bp1    bp2 num_cells_expressed overlap                  type</span>
<span class="token comment" spellcheck="true"># chr18_10025_10225     chr18_10025_10225  18  10025  10225                   5     201            Acetylated</span>
<span class="token comment" spellcheck="true"># chr18_10603_11103     chr18_10603_11103  18  10603  11103                   1  98,201 Acetylated,Methylated</span>
<span class="token comment" spellcheck="true"># chr18_11604_13986     chr18_11604_13986  18  11604  13986                   9      NA                  &lt;NA&gt;</span>
<span class="token comment" spellcheck="true"># chr18_49557_50057     chr18_49557_50057  18  49557  50057                   2      58            Acetylated</span>
<span class="token comment" spellcheck="true"># chr18_50240_50740     chr18_50240_50740  18  50240  50740                   2     501            Acetylated</span>
<span class="token comment" spellcheck="true"># chr18_104385_104585 chr18_104385_104585  18 104385 104585                   1     201            Methylated</span></code></pre></figure>
      <p>
        As you see above, when <code>all</code> is set to <code>TRUE</code>, all of the intervals that overlap are 
        reported.
      </p>
      <h3 id="find-overlapping-coordinates"><code>find_overlapping_coordinates</code></h3>
      <p>
        Lastly, you may simply want to know which of your peaks overlap a specific region of the genome. For that 
        Cicero includes the function <code>find_overlapping_coordinates</code>. 
      </p>
      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">find_overlapping_coordinates</span><span class="token punctuation">(</span><span class="token function">fData</span><span class="token punctuation">(</span><span class="token variable">temp</span><span class="token punctuation">)</span><span class="token punctuation">$</span><span class="token property">site_name</span><span class="token punctuation">,</span> <span class="token string">"chr18:10,100-10,604"</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># [1] "chr18_10025_10225" "chr18_10603_11103"</span>
      </code></pre></figure>


      <h2 id="frequently-asked-questions">Frequently Asked Questions </h2>

      <h4>Where do I get a genome_coordinates file for my genome/build?</h4>
      <p>
        The genome_coordinates files are simply a table for chromosomes and their sizes. You can find one
        for your genome/build on the USCS genome browser website. Go <a href="http://hgdownload.soe.ucsc.edu/downloads.html">here</a>,
        find the "Full Data Set" link for your genome of interest, and find the [genome].chrom.sizes file. Download this and pass 
        the path into the cicero function you're running! Note: some genomes have a lot of non-standard chromosomes in the sizes file 
        (chr_unk, alternative haplotypes, etc). You may want to remove these lines from your file, which will help speed up your run.
      </p>

      <h4>Where do I get the gene_model/gene_annotation_sample for my genome/build?</h4>
      <p>
        A gene_model to pass to plotting functions can be derived from the ENSEMBL GTF for your
        organism/build. Download the GTF file from <a href="https://uswest.ensembl.org/info/data/ftp/index.html">here</a> - found
        under the "Gene sets" column. GTF files need a bit of processing to be ready for plotting:
</p><figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token comment" spellcheck="true"># load in your data using rtracklayer</span>
<span class="token variable">gene_anno</span> <span class="token operator">&lt;-</span> <span class="token namespace">rtracklayer</span><span class="token punctuation">:</span><span class="token punctuation">:</span><span class="token function">readGFF</span><span class="token punctuation">(</span><span class="token string">"Path/To/Downloaded/GTF.GTF"</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># rename some columns to match plotting requirements</span>
<span class="token variable">gene_anno</span><span class="token punctuation">$</span><span class="token property">chromosome</span> <span class="token operator">&lt;-</span> <span class="token function">paste0</span><span class="token punctuation">(</span><span class="token string">"chr"</span><span class="token punctuation">,</span> <span class="token variable">gene_anno</span><span class="token punctuation">$</span><span class="token property">seqid</span><span class="token punctuation">)</span>
<span class="token variable">gene_anno</span><span class="token punctuation">$</span><span class="token property">gene</span> <span class="token operator">&lt;-</span> <span class="token variable">gene_anno</span><span class="token punctuation">$</span><span class="token property">gene_id</span>
<span class="token variable">gene_anno</span><span class="token punctuation">$</span><span class="token property">transcript</span> <span class="token operator">&lt;-</span> <span class="token variable">gene_anno</span><span class="token punctuation">$</span><span class="token property">transcript_id</span>
<span class="token variable">gene_anno</span><span class="token punctuation">$</span><span class="token property">symbol</span> <span class="token operator">&lt;-</span> <span class="token variable">gene_anno</span><span class="token punctuation">$</span><span class="token property">gene_name</span></code></pre></figure>
        
      <p></p>

      <h2 id="citation">Citation </h2>

      If you use Cicero to analyze your experiments, please cite:

      <figure class="highlight"><pre class=" language-r"><code class=" language-r" data-lang="r"><span class="token function">citation</span><span class="token punctuation">(</span><span class="token string">"cicero"</span><span class="token punctuation">)</span>

<span class="token comment" spellcheck="true"># Hannah A. Pliner, Jay Shendure &amp; Cole Trapnell et. al. (2018). Cicero</span>
<span class="token comment" spellcheck="true"># Predicts cis-Regulatory DNA Interactions from Single-Cell Chromatin</span>
<span class="token comment" spellcheck="true"># Accessibility Data. Molecular Cell, 71, 858 - 871.e8.</span>

<span class="token comment" spellcheck="true"># A BibTeX entry for LaTeX users is</span>

<span class="token comment" spellcheck="true"># @Article{,</span>
<span class="token comment" spellcheck="true">#   title = {Cicero Predicts cis-Regulatory DNA Interactions from Single-Cell Chromatin Accessibility Data},</span>
<span class="token comment" spellcheck="true">#   journal = {Molecular Cell},</span>
<span class="token comment" spellcheck="true">#   volume = {71},</span>
<span class="token comment" spellcheck="true">#   number = {5},</span>
<span class="token comment" spellcheck="true">#   pages = {858 - 871.e8},</span>
<span class="token comment" spellcheck="true">#   year = {2018},</span>
<span class="token comment" spellcheck="true">#   issn = {1097-2765},</span>
<span class="token comment" spellcheck="true">#   doi = {https://doi.org/10.1016/j.molcel.2018.06.044},</span>
<span class="token comment" spellcheck="true">#   author = {Hannah A. Pliner and Jonathan S. Packer and Jos L. McFaline-Figueroa and </span>
<span class="token comment" spellcheck="true">#     Darren A. Cusanovich and Riza M. Daza and Delasa Aghamirzaie and Sanjay Srivatsan </span>
<span class="token comment" spellcheck="true">#     and Xiaojie Qiu and Dana Jackson and Anna Minkina and Andrew C. Adey and Frank J. </span>
<span class="token comment" spellcheck="true">#     Steemers and Jay Shendure and Cole Trapnell},</span>
<span class="token comment" spellcheck="true"># }</span>

      </code></pre></figure>

      <h2 id="acknowledgements">Acknowledgements </h2>

      Cicero was developed by Hannah Pliner in Cole Trapnell's and Jay Shendure's labs.
      <br>
      <br>
      This vignette was created from Wolfgang Huber's Bioconductor vignette style document, and patterned after the 
      vignette for <em>DESeq</em>, by Simon Anders and Wolfgang Huber.

      <h2 id="references">References </h2>
      <p>
        Blondel, V.D., Guillaume, J.-L., Lambiotte, R., and Lefebvre, E. (2008). 
        <a href="http://iopscience.iop.org/article/10.1088/1742-5468/2008/10/P10008/meta">Fast unfolding of 
          communities in large networks.</a>
      </p>
      <p>
        Dekker, J., Marti-Renom, M.A., and Mirny, L.A. (2013). <a href="https://www.nature.com/articles/nrg3454">
          Exploring the three-dimensional organization of genomes: interpreting chromatin interaction data</a>. Nat. 
        Rev. Genet. 14, 390403.
      </p>
      <p>
        Sanborn, A.L., Rao, S.S.P., Huang, S.-C., Durand, N.C., Huntley, M.H., Jewett, A.I., Bochkov, I.D., 
        Chinnappan, D., Cutkosky, A., Li, J., et al. (2015). <a href="http://www.pnas.org/content/112/47/E6456">
          Chromatin extrusion explains key features of loop and domain formation in wild-type and engineered genomes.
        </a>. Proc. Natl. Acad. Sci. U. S. A. 112, E6456E6465.
      </p>
      <p>
        Sexton, T., Yaffe, E., Kenigsberg, E., Bantignies, F., Leblanc, B., Hoichman, M., Parrinello, H., Tanay, A., 
        and Cavalli, G. (2012). <a href="https://www.sciencedirect.com/science/article/pii/S0092867412000165">
          Three-Dimensional Folding and Functional Organization Principles of the Drosophila Genome.
        </a>. Cell 148:3, 458-472.
      </p>
    </div>
  </div>
</div>


    <script src="./Cicero_files/jquery.min.js"></script>
<script src="./Cicero_files/bootstrap.min.js"></script>
<script src="./Cicero_files/prism.js"></script>
<script src="./Cicero_files/prism.r.js" type="text/javascript"></script>
<script src="./Cicero_files/bootstrap-toc.min.js"></script>

<script type="text/x-mathjax-config;executed=true">
  MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
</script>
<script type="text/javascript" async="" src="./Cicero_files/MathJax.js">
</script>

<footer>

  <div class="wrapper">
      <p> 2018 Cole Trapnell</p>
  </div>

</footer>


  


</body></html>